---
title: "Kidney scRNAseq"
author: "Hailey Hampson"
date: "2025-01-25"
output:
  word_document: default
  html_document: default
  pdf_document: default
---
# 1. Set up Libraries & Directores
```{r, include=F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
library(GSEABase)
library(limma)
library(reshape2)
library(data.table)
library(knitr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
#library(NMF)
library(rsvd)
library(RColorBrewer)
library(MAST)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
#library(ideas)
library(foreach)
library(parallel)
library(doRNG)
library(doParallel)
library(fs)
# registerDoParallel(cores = 6)
library(VennDiagram)
library(janitor)
# devtools::install_github('immunogenomics/presto')
library(presto)
library(knitr)
library(lme4)
library(lmerTest)
#install.packages("glmmTMB")
# Reinstall glmmTMB from source
#install.packages("glmmTMB", type = "source")
#library(glmmTMB)

#Set number of cores for parallellization
#maxCores <- detectCores()
#numCores <- maxCores-1
#cl <- makeCluster(numCores)  # Create a cluster with the desired number of cores
#registerDoParallel(cl) 

#Local file path
#dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
#dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
#dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
#dir.results <- c("/Volumes/Peds Endo/Petter Bjornstad/Kidney Project/Results")

# #Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Kidney scRNA/Kidney scRNA")
# dir.results <- c(fs::path(dir.dat,"Kidney Project/Results"))

#Mac Studio File Path
dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Kidney scRNAseq Project/Results")
dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Load functions
source("Kidney_functions_sc.R")
```

# 2. Load Full Data & Clean: scRNA & MetaData
## a. Format Data
```{r, include=F}
#Local
# dir.dat <- c("/Users/hhampson/Dropbox/Bjornstad data")
# so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
# load("/Volumes/Peds Endo/Petter Bjornstad/Data Harmonization/Data Exports/PB90_clinical_metadata.RData")
# so_kidney_sc <- AddMetaData(so_kidney_sc, so_meta_combined)
# rm(so_meta_combined)
# gc()

#Lambda
so_kidney_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="KI-0014643")] <- "KL-0014643"
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="kl-0023998")] <- "KL-0023998"
#check2 <- so_kidney_sc@meta.data

#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv")) #small meta set
#Lambda
# harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata2.csv")) #all metadata vars
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata.csv")) #all metadata vars
#Mac studio
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))
#Get coenrolled IDs
coenroll <- read.csv(fs::path(dir.dat,"Renal HERITAGE/Data_Cleaned/coenrolled_ids.csv"))

#Which IDs are coenrolled 
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")]
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")]

#"KL-0014634" "KL-0014629" "KL-0014628" coenrolled IDs
#"KL-0014634" "KL-0014629" "KL-0014628" having missing meta data 

#KL-0014628: IT_10/RH-66-T KL-0014628 KL-0019101
#KL-0014629: IT_09/RH-65-T
#KL-0014634: IT_07/RH-59-T
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group")))

#Partipant 30051 (kit id KL-0027474 at 4m has sc, but at baseline no sc has kit id KL-0027474 metadata)
#missing <- c("KL-0014628", "KL-0014629", "KL-0014634", "KL-0019059", "KL-0019092", "KL-0019101", "KL-0022159", "KL-0024005", "KL-0027470", #"KL-0027478")
#Check these ids in metadat
#check <- harm_meta_data %>%
#filter(kit_id %in% missing)

#IDs missing group information  
#Filter out attempt participants
harm_meta_data <- harm_meta_data %>% 
  dplyr::select(-X)
#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>% 
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>% 
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>% 
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
#mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups: 
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
gc()
harm_meta_data <- harm_meta_data %>% 
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
gc()
meta_kidney_sc <-  so_kidney_sc@meta.data 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
meta_kidney_sc <- meta_kidney_sc %>%
  #dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(harm_meta_data,by="kit_id")
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
#Check to see if all ids still have metadata
#check4 <- meta_kidney_sc %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check4 <- check4 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))

#rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
ids <- harm_meta_data$kit_id
#Filter seurat object to only IDs that have the metadata (83 individuals of interest)
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)
#check5 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check5 <- check5 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
so_kidney_sc <- subset(so_kidney_sc, kit_id %in% ids)
#check6 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check6 <- check6 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
#unique(so_kidney_sc$group)
#unique(so_kidney_sc$KPMP_celltype)
rm(meta_kidney_sc,harm_meta_data)

#Switch default assay in seurat object to RNA
DefaultAssay(so_kidney_sc) <- "RNA"
gc()

check3 <- so_kidney_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","celltype_rpca","medication")))
# rm(check3)
unique(so_kidney_sc$co_enroll_id) 
#"RH-66-T" "RH-65-T" "RH-59-T" "IT2D-08" "RH-98-T" "RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T"
#"RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T" IDs coenrolled in RH and RH2 as a baseline and followup - filter out their RH2 study visit?
#RH-67-T/RH2-19 kit id for second visit to filter out: KL-0030621
#RH-23/RH2-14 kit id for second visit to filter out: KL-0029535

# #Check if coenrolled IDs have multiple biopsies
# co1 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")] #Coenrolled in RH&RH2
# co2 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH/RH2
# co3 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH2
# co4 <- coenroll$rh_id[which(coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenroleld in Improve/RH
# co5 <- coenroll$crc_id[which(coenroll$crc_id!="" & coenroll$panda_id!="")] #Coenrolled in Crocodile and panda
# co6 <- coenroll$panda_id[which(coenroll$panda_id!="" & coenroll$crc_id!="")]
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co1)]) #"RH-68-T" "RH-23-T" "RH-60-T" "RH-67-T" "RH-93-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co2)]) #"RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co3)]) #"RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co4)]) #"RH-66-T" "RH-65-T" "RH-59-T" "RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co5)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co6)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# 
# rm(coenroll,check3)


so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0030621")
so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0029535")
check3 <- so_kidney_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","celltype_rpca","medication")))
length(unique(check3$kit_id)) #81
rm(check3)
rm(coenroll)

#Check the total number of cells & genes in the full dataset
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #31332 genes

dim(so_kidney_sc@assays$RNA@counts) #31332 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #101

# #Filter out mitochondrial DNA
# gene_expression_matrix <- so_kidney_sc@assays$RNA@counts
# 
# # Filter genes expressed in more than 10% of cells
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# so_kidney_sc@assays$RNA@counts <- gene_expression_matrix[expressed_genes,]
# 
# dim(so_kidney_sc@assays$RNA@counts) #5759 186125 
# sum(grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))) #45
# 
# ncol(so_kidney_sc) #186125 cells
# nrow(so_kidney_sc) #31332 genes
# 
# # Remove mitochondrial genes (prefix "MT")
# so_kidney_sc@assays$RNA@counts <- so_kidney_sc@assays$RNA@counts[grep("^MT", rownames(so_kidney_sc@assays$RNA@counts), invert = TRUE),]
# ncol(so_kidney_sc) #186125 cells
# nrow(so_kidney_sc) #31332 genes
# dim(so_kidney_sc@assays$RNA@counts) #5714 186125
# sum(grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))) #0

# Step 1: Filter to genes expressed in more than 5% of cells
gene_expression_matrix <- so_kidney_sc@assays$RNA@counts
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05

so_kidney_sc <- subset(so_kidney_sc, features = rownames(gene_expression_matrix)[expressed_genes])
dim(so_kidney_sc@assays$RNA@counts) #9289 186125
dim(so_kidney_sc@assays$RNA) #9289 186125
dim(so_kidney_sc) #9289 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #70
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #9289 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_kidney_sc), value = TRUE)
#keep_ids <- unique(rownames(so_kidney_sc)[which(!rownames(so_kidney_sc) %in% mito_genes)])
# so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc), mito_genes))
# so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0029535")
#so_kidney_sc$Gene <- rownames(so_kidney_sc)
#so_kidney_sc <- subset(so_kidney_sc, Gene %in% keep_ids)
so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc@assays$RNA@counts), mito_genes))
#so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc@assays$RNA@counts), mito_genes))
# so_kidney_sc <- subset(so_kidney_sc, !rownames(so_kidney_sc) %in% mito_genes)
# grep("^MT-", rownames(so_kidney_sc@assays$RNA@counts), value = TRUE)
dim(so_kidney_sc@assays$RNA@counts) #9276 186125
dim(so_kidney_sc@assays$RNA@data) #9276 186125
dim(so_kidney_sc@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #0
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #9276 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_kidney_sc@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_kidney_sc@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_kidney_sc <- so_kidney_sc[, cells_to_keep]

# Check dimensions after filtering
dim(so_kidney_sc@assays$RNA@counts) #9276 81882
dim(so_kidney_sc@assays$RNA)#9276 81882
dim(so_kidney_sc)#9276 81882

# Optionally, you can also check how many MT genes remain in the filtered dataset
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts)))
ncol(so_kidney_sc)  # Number of cells remaining


# Step 3: Renormalize the Seurat object after filtering
so_kidney_sc <- NormalizeData(so_kidney_sc)

#Check normalized slot
dim(so_kidney_sc@assays$RNA@data) #5714 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@data))) #70
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #5714 genes

#Load in med data to update medication information
#Mac studio
med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))
#lambda
# med <- read.xlsx(fs::path(dir.dat,"Data Harmonization/Biopsies/Biopsies_w_mrn_Oct3.xlsx"))
#/Volumes/Peds Endo/Petter Bjornstad/Data Harmonization/Biopsies
med <- med %>% 
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
meta_kidney_sc <-  so_kidney_sc@meta.data
med <- med %>% 
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain
med <- med %>%  
  filter(record_id %in% meta_kidney_sc$record_id) #81 remain
# med$mrn <- as.numeric(med$mrn)
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
med$mrn <- as.numeric(med$mrn)
meta_kidney_sc <- meta_kidney_sc %>%
  left_join(med,by=c("mrn","record_id"))
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)

#Add Med Meta Data to Seurat object
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)
rm(gene_expression_matrix,med,meta_kidney_sc)
# saveRDS(so_kidney_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds"))


#Create UMAP of single cell 
#Create UMAP by individual? Label cell clusters

#Pseudobulk by individual
# Check if 'kit_id' exists in the metadata
length(unique(so_kidney_sc$kit_id)) #81 unique kit ids

#Check proportion first 

#Create PT and TAL pseudobulk cell type variable
so_kidney_sc$celltype <- case_when(grepl("PT-",so_kidney_sc$celltype_rpca)~"PT",
                                   grepl("TAL-",so_kidney_sc$celltype_rpca)~"TAL",
                                   grepl("EC-",so_kidney_sc$celltype_rpca)~"EC",
                                   grepl("POD",so_kidney_sc$celltype_rpca)~"POD",
                                   grepl("MAC",so_kidney_sc$celltype_rpca)~"MAC",
                                   grepl("MON",so_kidney_sc$celltype_rpca)~"MON",
                                   grepl("PC-",so_kidney_sc$celltype_rpca)~"PC",
                                   grepl("FIB",so_kidney_sc$celltype_rpca)~"FIB_MC_VSMC",
                                   grepl("DTL",so_kidney_sc$celltype_rpca)~"DTL",
                                   so_kidney_sc$celltype_rpca=="DCT"~"DCT",
                                   so_kidney_sc$celltype_rpca=="ATL"~"ATL",
                                   so_kidney_sc$celltype_rpca=="B"~"B",
                                   so_kidney_sc$celltype_rpca=="T"~"T")
so_kidney_sc$celltype <- as.character(so_kidney_sc$celltype)

# # Aggregate the expression data by kit_id (or individual)
# # This sums the raw counts across all cells within each individual (kit_id)
# pseudobulked_data <- AggregateExpression(so_kidney_sc, group.by = "kit_id")
# 
# # Now pseudobulked_data contains summed counts for each gene by kit_id
# # Create a new Seurat object using the pseudobulked data
# so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA)
# 
# # Use GetAssayData to access the raw counts from the RNA assay
# # raw_counts <- GetAssayData(so_pb, assay = "RNA", layer = "counts")
# 
# so_pb <- NormalizeData(so_pb)
# length(unique(rownames(so_pb))) #9276 individual genes (rownames)
# length(unique(colnames(so_pb))) #81 individuals (colnames)
# 
# #Add origional metadata
# meta.data <- so_kidney_sc@meta.data
# #Get metadat of interest: age, sex, group, group2, meds, 
# vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c",
#           "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi",
#          "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1",
#          "mfm_1","insulin_1","medication","group","group2","acr_u","KPMP_celltype","celltype_rpca")
# aggregated_metadata <- meta.data %>%
#   group_by(kit_id) %>%
#   summarise(across(all_of(vars), first))
# aggregated_metadata <- as.data.frame(aggregated_metadata)
# rownames(aggregated_metadata) <- rownames(so_pb@meta.data)
# 
# #Add Med Meta Data to Seurat object
# so_pb <- AddMetaData(so_pb, aggregated_metadata)
# # meta <- so_pb@meta.data
# rm(aggregated_metadata,pseudobulked_data,meta.data)
# 
# # Aggregate expression by both 'celltype' and 'kit_id'
pseudobulked_data <- PseudobulkExpression(so_kidney_sc, method="average",group.by = c("celltype", "kit_id"),return.seurat=T)
ncol(pseudobulked_data) #992 cell/individuals
nrow(pseudobulked_data) #9276
pseudobulked_data_sum <- PseudobulkExpression(so_kidney_sc, method="aggregate",group.by = c("celltype", "kit_id"),return.seurat=T,normalization.metho="LogNormalize")
ncol(pseudobulked_data_sum) #992 cell/individuals
nrow(pseudobulked_data_sum) #9276

# Now pseudobulked_data contains summed counts for each gene by both celltype and kit_id
# Create a new Seurat object using the pseudobulked data
# so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA)
# ncol(so_pb) #162 individuals/cells
# nrow(so_pb) #9276 genes

# # Normalize the data
# so_pb <- NormalizeData(so_pb)

# # Check the number of unique genes and individuals (celltypes and kit_ids combined)
# length(unique(rownames(so_pb)))  # Number of individual genes (rows)
# length(unique(colnames(so_pb)))  # Number of unique combinations of celltype and kit_id (columns)

# Add original metadata
meta.data <- so_kidney_sc@meta.data

# Get metadata of interest: age, sex, group, etc.
vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c",
          "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi",
          "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1",
          "mfm_1","insulin_1","medication","group","group2","acr_u","celltype_rpca")

# Aggregate metadata by both 'celltype' and 'kit_id'
aggregated_metadata <- meta.data %>%
  filter(!is.na(celltype)) %>%
  group_by(celltype, kit_id) %>%
  summarise(across(all_of(vars), first))

# # Count the number of cells in each celltype and kit_id
# # Extract metadata from the Seurat object
# cell_count_per_group <- meta.data %>%
#   filter(!is.na(celltype)) %>%
#   group_by(celltype, group) %>%
#   summarize(cell_count = n())
# 
# # Calculate total number of cells in each kit_id group
# total_cells_per_kit <- meta.data %>%
#   filter(!is.na(celltype)) %>%
#   group_by(group) %>%
#   summarize(total_cells = n())
# 
# # Merge cell count with total cells per kit_id to get the proportions
# proportion_cells <- cell_count_per_group %>%
#   left_join(total_cells_per_kit, by = "group") %>%
#   mutate(proportion = cell_count / total_cells)
# 
# ggplot(proportion_cells, aes(x = group, y = proportion, fill = celltype)) +
#   geom_bar(stat = "identity", position = "stack") +
#   theme_minimal() +
#   labs(title = "Proportion of Cells Pseudobulked by Cell Type and Disease Group",
#        x = "Disease Status",
#        y = "Proportion of Cells") +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Convert to a data frame and set the rownames
aggregated_metadata <- as.data.frame(aggregated_metadata)
rownames(aggregated_metadata) <- paste(aggregated_metadata$celltype, aggregated_metadata$kit_id, sep = "_")

# Add the aggregated metadata to the Seurat object
# so_pb <- AddMetaData(so_pb, aggregated_metadata)
pseudobulked_data <- AddMetaData(pseudobulked_data, aggregated_metadata)
pseudobulked_data_sum <- AddMetaData(pseudobulked_data_sum, aggregated_metadata)


# Clean up unnecessary objects
rm(aggregated_metadata, meta.data)
# so_pb_total <- so_pb
# rm(so_pb)

# Extract gene expression data (normalized counts, for example)
expression_data <- GetAssayData(pseudobulked_data, layer = "data")

# Convert gene expression data to a data frame
# Transpose it if you want genes as rows and cells as columns
expression_df <- as.data.frame(t(expression_data))
# Extract metadata
metadata_df <- pseudobulked_data@meta.data
# Optional: Combine both into a single data frame
combined_df <- cbind(metadata_df, expression_df)
data_pb <- combined_df

# Extract gene expression data (normalized counts, for example)
expression_data <- GetAssayData(pseudobulked_data_sum, layer = "data")

# Convert gene expression data to a data frame
# Transpose it if you want genes as rows and cells as columns
expression_df <- as.data.frame(t(expression_data))
# Extract metadata
metadata_df <- pseudobulked_data_sum@meta.data
# Optional: Combine both into a single data frame
combined_df <- cbind(metadata_df, expression_df)
data_pb_sum <- combined_df
rm(combined_df,metadata_df,expression_df,expression_data,pseudobulked_data,pseudobulked_data_sum)
```

<!-- ## b. UMAPs -->
<!-- ```{r, echo=F, warning=F, fig.width=15,fig,height=10} -->
<!-- #Create UMAP -->
<!-- dim(so_kidney_sc) #9276 genes, 3265 cells -->

<!-- #Umap visualizeations -->
<!-- # Step 3: Renormalize the Seurat object after filtering -->
<!-- so_kidney_sc <- ScaleData(so_kidney_sc) -->
<!-- # PCA -->
<!-- # so_kidney_sc@reductions$umap <- NULL -->
<!-- # so_kidney_sc@meta.data$umap_ <- NULL -->
<!-- so_kidney_sc <- RunPCA(so_kidney_sc, features = VariableFeatures(object = so_kidney_sc)) -->
<!-- ElbowPlot(so_kidney_sc) -->
<!-- # Cluster cells -->
<!-- so_kidney_sc <- FindNeighbors(so_kidney_sc) -->
<!-- so_kidney_sc <- FindClusters(so_kidney_sc) -->
<!-- # Perform UMAP and tSNE -->
<!-- # so_kidney_sc@reductions$umap <- NULL -->
<!-- # so_kidney_sc@meta.data$umap_ <- NULL -->
<!-- so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15) -->
<!-- DimPlot(so_kidney_sc, reduction = "umap.rpca")  -->

<!-- DimPlot(so_kidney_sc, reduction = "umap.rpca") + -->
<!--   labs(x = "umap_1", y = "umap_2") + -->
<!--   ggtitle("UMAP") -->

<!-- # Plot UMAP and label by cell type stored in KPMP_celltype -->
<!-- DimPlot(so_kidney_sc, reduction = "umap.rpca", group.by = "celltype_rpca", label = TRUE) + -->
<!--   ggtitle("UMAP by Cell Type") + -->
<!--   labs(x = "umap_1", y = "umap_2")  -->




<!-- ``` -->

<!-- ## c. Format KPMP Data -->
<!-- ```{r, include=F} -->
<!-- #Load KPMP Cell Types 2 (new) -->
<!-- #Mac Studio -->
<!-- so_kpmp_sc <- readRDS("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/scRNA/data_raw/PB_90_RPCAFix_Metadata_LR_RM_kpmpV1labelled.rds") -->

<!-- so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="KI-0014643")] <- "KL-0014643" -->
<!-- so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="kl-0023998")] <- "KL-0023998" -->
<!-- #check2 <- so_kpmp_sc@meta.data -->

<!-- #harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv")) #small meta set -->
<!-- #harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata.csv")) #all metadata vars -->
<!-- harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv")) -->
<!-- #Get coenrolled IDs -->
<!-- coenroll <- read.csv(fs::path(dir.dat,"Renal HERITAGE/Data_Cleaned/coenrolled_ids.csv")) -->

<!-- #Which IDs are coenrolled -->
<!-- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")] -->
<!-- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")] -->

<!-- #"KL-0014634" "KL-0014629" "KL-0014628" coenrolled IDs -->
<!-- #"KL-0014634" "KL-0014629" "KL-0014628" having missing meta data -->

<!-- #KL-0014628: IT_10/RH-66-T KL-0014628 KL-0019101 -->
<!-- #KL-0014629: IT_09/RH-65-T -->
<!-- #KL-0014634: IT_07/RH-59-T -->
<!-- #check <- harm_meta_data %>% -->
<!-- #  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group"))) -->

<!-- #Partipant 30051 (kit id KL-0027474 at 4m has sc, but at baseline no sc has kit id KL-0027474 metadata) -->
<!-- #missing <- c("KL-0014628", "KL-0014629", "KL-0014634", "KL-0019059", "KL-0019092", "KL-0019101", "KL-0022159", "KL-0024005", "KL-0027470", #"KL-0027478") -->
<!-- #Check these ids in metadat -->
<!-- #check <- harm_meta_data %>% -->
<!-- #filter(kit_id %in% missing) -->

<!-- #IDs missing group information -->
<!-- #Filter out attempt participants -->
<!-- harm_meta_data <- harm_meta_data %>% -->
<!--   dplyr::select(-X) -->
<!-- #Create Medication Groups -->
<!-- harm_meta_data <- harm_meta_data %>% -->
<!--   mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>% -->
<!--   mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>% -->
<!--   mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>% -->
<!--   mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No")) -->
<!-- #mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No")) -->

<!-- #Define 4 exposure groups: -->
<!-- #SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-) -->
<!-- gc() -->
<!-- harm_meta_data <- harm_meta_data %>% -->
<!--   mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2", -->
<!--                                 sglt2 == "Yes" ~ "sglt2", -->
<!--                                 glp1 == "Yes" ~ "glp1", -->
<!--                                 no_med == "Yes" ~ "no_med")) -->
<!-- harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #Define a T2D+OB group -->
<!-- harm_meta_data <- harm_meta_data %>% -->
<!--   mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group)) -->
<!-- gc() -->
<!-- meta_kidney_sc <-  so_kpmp_sc@meta.data -->
<!-- rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data) -->
<!-- meta_kidney_sc <- meta_kidney_sc %>% -->
<!--   #dplyr::mutate(RNAlater_ID = SampleID) %>% -->
<!--   left_join(harm_meta_data,by="kit_id") -->
<!-- rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data) -->
<!-- #Check to see if all ids still have metadata -->
<!-- #check4 <- meta_kidney_sc %>% -->
<!-- #  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634") -->
<!-- #check4 <- check4 %>% -->
<!-- #  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype"))) -->

<!-- #rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data) -->
<!-- ids <- harm_meta_data$kit_id -->
<!-- #Filter seurat object to only IDs that have the metadata (83 individuals of interest) -->
<!-- so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc) -->
<!-- #check5 <- so_kpmp_sc@meta.data %>% -->
<!-- #  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634") -->
<!-- #check5 <- check5 %>% -->
<!-- #  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype"))) -->
<!-- so_kpmp_sc <- subset(so_kpmp_sc, kit_id %in% ids) -->
<!-- #check6 <- so_kpmp_sc@meta.data %>% -->
<!-- #  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634") -->
<!-- #check6 <- check6 %>% -->
<!-- #  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype"))) -->
<!-- #unique(so_kpmp_sc$group) -->
<!-- #unique(so_kpmp_sc$KPMP_celltype) -->
<!-- rm(meta_kidney_sc,harm_meta_data) -->

<!-- #Switch default assay in seurat object to RNA -->
<!-- DefaultAssay(so_kpmp_sc) <- "RNA" -->
<!-- gc() -->

<!-- check3 <- so_kpmp_sc@meta.data -->
<!-- check3 <- check3 %>% -->
<!--   dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","KPMP_celltype","medication"))) -->
<!-- # rm(check3) -->
<!-- unique(so_kpmp_sc$co_enroll_id) -->
<!-- length(unique(check3$kit_id)) -->
<!-- #"RH-66-T" "RH-65-T" "RH-59-T" "IT2D-08" "RH-98-T" "RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T" -->
<!-- #"RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T" IDs coenrolled in RH and RH2 as a baseline and followup - filter out their RH2 study visit? -->
<!-- #RH-67-T/RH2-19 kit id for second visit to filter out: KL-0030621 -->
<!-- #RH-23/RH2-14 kit id for second visit to filter out: KL-0029535 -->

<!-- # #Check if coenrolled IDs have multiple biopsies -->
<!-- # co1 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")] #Coenrolled in RH&RH2 -->
<!-- # co2 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH/RH2 -->
<!-- # co3 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH2 -->
<!-- # co4 <- coenroll$rh_id[which(coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenroleld in Improve/RH -->
<!-- # co5 <- coenroll$crc_id[which(coenroll$crc_id!="" & coenroll$panda_id!="")] #Coenrolled in Crocodile and panda -->
<!-- # co6 <- coenroll$panda_id[which(coenroll$panda_id!="" & coenroll$crc_id!="")] -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co1)]) #"RH-68-T" "RH-23-T" "RH-60-T" "RH-67-T" "RH-93-T" -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co2)]) #"RH-60-T" -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co3)]) #"RH-60-T" -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co4)]) #"RH-66-T" "RH-65-T" "RH-59-T" "RH-60-T" -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co5)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59" -->
<!-- # unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co6)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59" -->
<!-- # -->
<!-- # rm(coenroll,check3) -->


<!-- so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0030621") -->
<!-- so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0029535") -->
<!-- check3 <- so_kpmp_sc@meta.data -->
<!-- check3 <- check3 %>% -->
<!--   dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","KPMP_celltype","medication"))) -->
<!-- length(unique(check3$kit_id)) #81 -->
<!-- rm(check3) -->
<!-- rm(coenroll) -->

<!-- #Check the total number of cells & genes in the full dataset -->
<!-- ncol(so_kpmp_sc) #186125 cells -->
<!-- nrow(so_kpmp_sc) #31332 genes -->

<!-- dim(so_kpmp_sc@assays$RNA@layers$counts) #31332 186125 -->
<!-- sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA))) #101 -->

<!-- # #Filter out mitochondrial DNA -->
<!-- # gene_expression_matrix <- so_kpmp_sc@assays$RNA@counts -->
<!-- # -->
<!-- # # Filter genes expressed in more than 10% of cells -->
<!-- # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1 -->
<!-- # so_kpmp_sc@assays$RNA@counts <- gene_expression_matrix[expressed_genes,] -->
<!-- # -->
<!-- # dim(so_kpmp_sc@assays$RNA@counts) #5759 186125 -->
<!-- # sum(grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts))) #45 -->
<!-- # -->
<!-- # ncol(so_kpmp_sc) #186125 cells -->
<!-- # nrow(so_kpmp_sc) #31332 genes -->
<!-- # -->
<!-- # # Remove mitochondrial genes (prefix "MT") -->
<!-- # so_kpmp_sc@assays$RNA@counts <- so_kpmp_sc@assays$RNA@counts[grep("^MT", rownames(so_kpmp_sc@assays$RNA@counts), invert = TRUE),] -->
<!-- # ncol(so_kpmp_sc) #186125 cells -->
<!-- # nrow(so_kpmp_sc) #31332 genes -->
<!-- # dim(so_kpmp_sc@assays$RNA@counts) #5714 186125 -->
<!-- # sum(grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts))) #0 -->

<!-- # Step 1: Filter to genes expressed in more than 5% of cells -->
<!-- gene_expression_matrix <- so_kpmp_sc@assays$RNA@layers$counts -->
<!-- rownames(gene_expression_matrix) <- rownames(so_kpmp_sc@assays$RNA) -->
<!-- colnames(gene_expression_matrix) <- colnames(so_kpmp_sc@assays$RNA) -->
<!-- # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1 -->
<!-- expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05 -->
<!-- names(expressed_genes) -->
<!-- rownames(gene_expression_matrix)[expressed_genes] -->

<!-- so_kpmp_sc <- subset(so_kpmp_sc, features = rownames(gene_expression_matrix)[expressed_genes]) -->
<!-- dim(so_kpmp_sc@assays$RNA@layers$counts) #9289 186125 -->
<!-- dim(so_kpmp_sc@assays$RNA) #9289 186125 -->
<!-- dim(so_kpmp_sc) #9289 186125 -->
<!-- sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA))) #70 -->
<!-- ncol(so_kpmp_sc) #186125 cells -->
<!-- nrow(so_kpmp_sc) #9289 genes -->

<!-- # Step 2: Remove mitochondrial genes (those starting with "MT") -->
<!-- mito_genes <- grep("^MT-", rownames(so_kpmp_sc@assays$RNA), value = TRUE) -->
<!-- so_kpmp_sc <- subset(so_kpmp_sc, features = setdiff(rownames(so_kpmp_sc), mito_genes)) -->
<!-- dim(so_kpmp_sc@assays$RNA@layers$counts) #9276 186125 -->
<!-- dim(so_kpmp_sc@assays$RNA@layers$data) #9276 186125 -->
<!-- dim(so_kpmp_sc@assays$RNA)#5714 186125 -->
<!-- sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$counts))) #0 -->
<!-- ncol(so_kpmp_sc) #186125 cells -->
<!-- nrow(so_kpmp_sc) #9276 genes -->

<!-- # # Identify mitochondrial genes -->
<!-- # mt_genes <- grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts)) -->
<!-- # -->
<!-- # # Calculate the percentage of mitochondrial genes per cell -->
<!-- # mt_percentage <- Matrix::colSums(so_kpmp_sc@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_kpmp_sc@assays$RNA@counts) -->
<!-- # -->
<!-- # # Filter out cells where the proportion of mitochondrial genes is > 15% -->
<!-- # cells_to_keep <- mt_percentage <= 0.15 -->
<!-- # -->
<!-- # # Subset the object to retain only cells with <= 15% mitochondrial genes -->
<!-- # so_kpmp_sc <- so_kpmp_sc[, cells_to_keep] -->

<!-- # Check dimensions after filtering -->
<!-- dim(so_kpmp_sc@assays$RNA@layers$counts) #9289 81882 -->
<!-- dim(so_kpmp_sc@assays$RNA)#9289 81882 -->
<!-- dim(so_kpmp_sc)#9289 81882 -->

<!-- # Optionally, you can also check how many MT genes remain in the filtered dataset -->
<!-- sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$counts))) -->
<!-- ncol(so_kpmp_sc)  # Number of cells remaining -->


<!-- # Step 3: Renormalize the Seurat object after filtering -->
<!-- so_kpmp_sc <- NormalizeData(so_kpmp_sc) -->

<!-- #Check normalized slot -->
<!-- dim(so_kpmp_sc@assays$RNA@layers$data) #5714 186125 -->
<!-- sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$data))) #70 -->
<!-- ncol(so_kpmp_sc) #186125 cells -->
<!-- nrow(so_kpmp_sc) #5714 genes -->

<!-- #Load in med data to update medication information -->
<!-- med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx")) -->
<!-- med <- med %>% -->
<!--   dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1"))) -->
<!-- meta_kidney_sc <-  so_kpmp_sc@meta.data -->
<!-- med <- med %>% -->
<!--   filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain -->
<!-- med <- med %>% -->
<!--   filter(record_id %in% meta_kidney_sc$record_id) #81 remain -->
<!-- # med$mrn <- as.numeric(med$mrn) -->
<!-- rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data) -->
<!-- med$mrn <- as.numeric(med$mrn) -->
<!-- meta_kidney_sc <- meta_kidney_sc %>% -->
<!--   left_join(med,by=c("mrn","record_id")) -->
<!-- rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data) -->

<!-- #Add Med Meta Data to Seurat object -->
<!-- so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc) -->
<!-- rm(gene_expression_matrix,med,meta_kidney_sc) -->
<!-- rm(expressed_genes) -->
<!-- # saveRDS(so_kpmp_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds")) -->

<!-- #Pseudobulk by individual -->
<!-- # Check if 'kit_id' exists in the metadata -->
<!-- length(unique(so_kpmp_sc$kit_id)) #81 unique kit ids -->

<!-- # #Create PT and TAL pseudobulk cell type variable -->
<!-- # so_kpmp_sc$celltype <- ifelse(grepl("PT-",so_kpmp_sc$celltype_rpca),"PT", -->
<!-- #                               ifelse(grepl("TAL-",so_kpmp_sc$celltype_rpca),"TAL",NA)) -->
<!-- # so_kpmp_sc$celltype <- as.character(so_kpmp_sc$celltype) -->
<!-- #  -->
<!-- # # # Aggregate the expression data by kit_id (or individual) -->
<!-- # # # This sums the raw counts across all cells within each individual (kit_id) -->
<!-- # # pseudobulked_data <- AggregateExpression(so_kpmp_sc, group.by = "kit_id") -->
<!-- # # -->
<!-- # # # Now pseudobulked_data contains summed counts for each gene by kit_id -->
<!-- # # # Create a new Seurat object using the pseudobulked data -->
<!-- # # so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA) -->
<!-- # # -->
<!-- # # # Use GetAssayData to access the raw counts from the RNA assay -->
<!-- # # # raw_counts <- GetAssayData(so_pb, assay = "RNA", layer = "counts") -->
<!-- # # -->
<!-- # # so_pb <- NormalizeData(so_pb) -->
<!-- # # length(unique(rownames(so_pb))) #9276 individual genes (rownames) -->
<!-- # # length(unique(colnames(so_pb))) #81 individuals (colnames) -->
<!-- # # -->
<!-- # # #Add origional metadata -->
<!-- # # meta.data <- so_kpmp_sc@meta.data -->
<!-- # # #Get metadat of interest: age, sex, group, group2, meds, -->
<!-- # # vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c", -->
<!-- # #           "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi", -->
<!-- # #          "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1", -->
<!-- # #          "mfm_1","insulin_1","medication","group","group2","acr_u","KPMP_celltype","celltype_rpca") -->
<!-- # # aggregated_metadata <- meta.data %>% -->
<!-- # #   group_by(kit_id) %>% -->
<!-- # #   summarise(across(all_of(vars), first)) -->
<!-- # # aggregated_metadata <- as.data.frame(aggregated_metadata) -->
<!-- # # rownames(aggregated_metadata) <- rownames(so_pb@meta.data) -->
<!-- # # -->
<!-- # # #Add Med Meta Data to Seurat object -->
<!-- # # so_pb <- AddMetaData(so_pb, aggregated_metadata) -->
<!-- # # # meta <- so_pb@meta.data -->
<!-- # # rm(aggregated_metadata,pseudobulked_data,meta.data) -->
<!-- #  -->
<!-- # # Aggregate expression by both 'celltype' and 'kit_id' -->
<!-- # pseudobulked_data <- AggregateExpression(so_kpmp_sc, group.by = c("celltype", "kit_id")) -->
<!-- #  -->
<!-- # # Now pseudobulked_data contains summed counts for each gene by both celltype and kit_id -->
<!-- # # Create a new Seurat object using the pseudobulked data -->
<!-- # so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA) -->
<!-- # ncol(so_pb) #162 individuals/cells -->
<!-- # nrow(so_pb) #9276 genes -->
<!-- #  -->
<!-- # # Normalize the data -->
<!-- # so_pb <- NormalizeData(so_pb) -->
<!-- #  -->
<!-- # # Check the number of unique genes and individuals (celltypes and kit_ids combined) -->
<!-- # length(unique(rownames(so_pb)))  # Number of individual genes (rows) -->
<!-- # length(unique(colnames(so_pb)))  # Number of unique combinations of celltype and kit_id (columns) -->
<!-- #  -->
<!-- # # Add original metadata -->
<!-- # meta.data <- so_kpmp_sc@meta.data -->
<!-- #  -->
<!-- # # Get metadata of interest: age, sex, group, etc. -->
<!-- # vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c", -->
<!-- #           "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi", -->
<!-- #           "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1", -->
<!-- #           "mfm_1","insulin_1","medication","group","group2","acr_u","KPMP_celltype","celltype_rpca") -->
<!-- #  -->
<!-- # # Aggregate metadata by both 'celltype' and 'kit_id' -->
<!-- # aggregated_metadata <- meta.data %>% -->
<!-- #   filter(!is.na(celltype)) %>% -->
<!-- #   group_by(celltype, kit_id) %>% -->
<!-- #   summarise(across(all_of(vars), first)) -->
<!-- #  -->
<!-- # # Convert to a data frame and set the rownames -->
<!-- # aggregated_metadata <- as.data.frame(aggregated_metadata) -->
<!-- # rownames(aggregated_metadata) <- paste(aggregated_metadata$celltype, aggregated_metadata$kit_id, sep = "_") -->
<!-- #  -->
<!-- # # Add the aggregated metadata to the Seurat object -->
<!-- # so_pb <- AddMetaData(so_pb, aggregated_metadata) -->
<!-- #  -->
<!-- # # Clean up unnecessary objects -->
<!-- # rm(aggregated_metadata, pseudobulked_data, meta.data) -->
<!-- # so_pb_total <- so_pb -->
<!-- # rm(so_pb) -->


<!-- #Create UMAP of single cell  -->
<!-- #Create UMAP by individual? Label cell clusters -->

<!-- #Pseudobulk by individual -->
<!-- # Check if 'kit_id' exists in the metadata -->
<!-- length(unique(so_kpmp_sc$kit_id)) #81 unique kit ids -->

<!-- #Check proportion first  -->

<!-- #Create PT and TAL pseudobulk cell type variable -->
<!-- so_kpmp_sc$celltype <- case_when(grepl("PT-",so_kpmp_sc$celltype_rpca)~"PT", -->
<!--                                  grepl("TAL-",so_kpmp_sc$celltype_rpca)~"TAL", -->
<!--                                  grepl("EC-",so_kpmp_sc$celltype_rpca)~"EC", -->
<!--                                  grepl("POD",so_kpmp_sc$celltype_rpca)~"POD", -->
<!--                                  grepl("MAC",so_kpmp_sc$celltype_rpca)~"MAC", -->
<!--                                  grepl("MON",so_kpmp_sc$celltype_rpca)~"MON", -->
<!--                                  grepl("PC-",so_kpmp_sc$celltype_rpca)~"PC", -->
<!--                                  grepl("FIB",so_kpmp_sc$celltype_rpca)~"FIB_MC_VSMC", -->
<!--                                  grepl("DTL",so_kpmp_sc$celltype_rpca)~"DTL", -->
<!--                                  so_kpmp_sc$celltype_rpca=="DCT"~"DCT", -->
<!--                                  so_kpmp_sc$celltype_rpca=="ATL"~"ATL", -->
<!--                                  so_kpmp_sc$celltype_rpca=="B"~"B", -->
<!--                                  so_kpmp_sc$celltype_rpca=="T"~"T") -->
<!-- so_kpmp_sc$celltype <- as.character(so_kpmp_sc$celltype) -->

<!-- # # Aggregate the expression data by kit_id (or individual) -->
<!-- # # This sums the raw counts across all cells within each individual (kit_id) -->
<!-- # pseudobulked_data <- AggregateExpression(so_kpmp_sc, group.by = "kit_id") -->
<!-- #  -->
<!-- # # Now pseudobulked_data contains summed counts for each gene by kit_id -->
<!-- # # Create a new Seurat object using the pseudobulked data -->
<!-- # so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA) -->
<!-- #  -->
<!-- # # Use GetAssayData to access the raw counts from the RNA assay -->
<!-- # # raw_counts <- GetAssayData(so_pb, assay = "RNA", layer = "counts") -->
<!-- #  -->
<!-- # so_pb <- NormalizeData(so_pb) -->
<!-- # length(unique(rownames(so_pb))) #9276 individual genes (rownames) -->
<!-- # length(unique(colnames(so_pb))) #81 individuals (colnames) -->
<!-- #  -->
<!-- # #Add origional metadata -->
<!-- # meta.data <- so_kpmp_sc@meta.data -->
<!-- # #Get metadat of interest: age, sex, group, group2, meds,  -->
<!-- # vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c", -->
<!-- #           "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi", -->
<!-- #          "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1", -->
<!-- #          "mfm_1","insulin_1","medication","group","group2","acr_u","KPMP_celltype","celltype_rpca") -->
<!-- # aggregated_metadata <- meta.data %>% -->
<!-- #   group_by(kit_id) %>% -->
<!-- #   summarise(across(all_of(vars), first)) -->
<!-- # aggregated_metadata <- as.data.frame(aggregated_metadata) -->
<!-- # rownames(aggregated_metadata) <- rownames(so_pb@meta.data) -->
<!-- #  -->
<!-- # #Add Med Meta Data to Seurat object -->
<!-- # so_pb <- AddMetaData(so_pb, aggregated_metadata) -->
<!-- # # meta <- so_pb@meta.data -->
<!-- # rm(aggregated_metadata,pseudobulked_data,meta.data) -->

<!-- # Aggregate expression by both 'celltype' and 'kit_id' -->
<!-- pseudobulked_data <- AggregateExpression(so_kpmp_sc, group.by = c("celltype", "kit_id")) -->

<!-- # Now pseudobulked_data contains summed counts for each gene by both celltype and kit_id -->
<!-- # Create a new Seurat object using the pseudobulked data -->
<!-- so_pb <- CreateSeuratObject(counts = pseudobulked_data$RNA) -->
<!-- ncol(so_pb) #162 individuals/cells -->
<!-- nrow(so_pb) #9276 genes -->

<!-- # Normalize the data -->
<!-- so_pb <- NormalizeData(so_pb) -->

<!-- # Check the number of unique genes and individuals (celltypes and kit_ids combined) -->
<!-- length(unique(rownames(so_pb)))  # Number of individual genes (rows) -->
<!-- length(unique(colnames(so_pb)))  # Number of unique combinations of celltype and kit_id (columns) -->

<!-- # Add original metadata -->
<!-- meta.data <- so_kpmp_sc@meta.data -->

<!-- # Get metadata of interest: age, sex, group, etc. -->
<!-- vars <- c("age","sex","bmi","dexa_fat_kg","dexa_lean_kg","hba1c", -->
<!--           "sbp","dbp","map","triglycerides","gfr_raw_plasma","eGFR_CKD_epi", -->
<!--           "erpf_raw_plasma","total_kidney_volume_ml","ht_adj_tkv","raasi_1", -->
<!--           "mfm_1","insulin_1","medication","group","group2","acr_u","celltype_rpca") -->

<!-- # Aggregate metadata by both 'celltype' and 'kit_id' -->
<!-- aggregated_metadata <- meta.data %>% -->
<!--   filter(!is.na(celltype)) %>% -->
<!--   group_by(celltype, kit_id) %>% -->
<!--   summarise(across(all_of(vars), first)) -->

<!-- # Convert to a data frame and set the rownames -->
<!-- aggregated_metadata <- as.data.frame(aggregated_metadata) -->
<!-- rownames(aggregated_metadata) <- paste(aggregated_metadata$celltype, aggregated_metadata$kit_id, sep = "_") -->

<!-- # Add the aggregated metadata to the Seurat object -->
<!-- so_pb <- AddMetaData(so_pb, aggregated_metadata) -->

<!-- # Clean up unnecessary objects -->
<!-- rm(aggregated_metadata, pseudobulked_data, meta.data) -->
<!-- so_pb_total <- so_pb -->
<!-- rm(so_pb) -->
<!-- ``` -->

<!-- ## d. KPMP UMAPs -->
<!-- ```{r, echo=F, warning=F, fig.width=15,fig,height=10} -->
<!-- #Create UMAP -->
<!-- dim(so_kpmp_sc) #9276 genes, 3265 cells -->

<!-- #Umap visualizeations -->
<!-- # Step 3: Renormalize the Seurat object after filtering -->
<!-- so_kpmp_sc <- ScaleData(so_kpmp_sc) -->
<!-- # PCA -->
<!-- # so_kpmp_sc@reductions$umap <- NULL -->
<!-- # so_kpmp_sc@meta.data$umap_ <- NULL -->
<!-- so_kpmp_sc <- RunPCA(so_kpmp_sc, features = VariableFeatures(object = so_kpmp_sc)) -->
<!-- ElbowPlot(so_kpmp_sc) -->
<!-- # Cluster cells -->
<!-- so_kpmp_sc <- FindNeighbors(so_kpmp_sc) -->
<!-- so_kpmp_sc <- FindClusters(so_kpmp_sc) -->
<!-- # Perform UMAP and tSNE -->
<!-- # so_kpmp_sc@reductions$umap <- NULL -->
<!-- # so_kpmp_sc@meta.data$umap_ <- NULL -->
<!-- so_kpmp_sc <- RunUMAP(so_kpmp_sc, dims = 1:15) -->
<!-- # DimPlot(so_kpmp_sc, reduction = "umap.rpca")  -->
<!-- #  -->
<!-- # DimPlot(so_kpmp_sc, reduction = "umap.rpca") + -->
<!-- #   labs(x = "umap_1", y = "umap_2") + -->
<!-- #   ggtitle("UMAP") -->

<!-- # Plot UMAP and label by cell type stored in KPMP_celltype -->
<!-- DimPlot(so_kpmp_sc, reduction = "umap.rpca", group.by = "KPMP_celltype", label = TRUE,raster=F) + -->
<!--   ggtitle("UMAP by KPMP Cell Type") + -->
<!--   labs(x = "umap_1", y = "umap_2")  -->

<!-- # Plot UMAP and label by cell type stored in KPMP_celltype -->
<!-- DimPlot(so_kpmp_sc, reduction = "umap.rpca", group.by = "celltype_rpca", label = TRUE,raster=F) + -->
<!--   ggtitle("UMAP by Cell Type") + -->
<!--   labs(x = "umap_1", y = "umap_2")  -->

<!-- #Type 2 diabetes -->
<!-- so_diab <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_diab <- NormalizeData(so_diab) -->
<!-- so_diab <- ScaleData(so_diab) -->
<!-- so_diab <- RunPCA(so_diab, features = VariableFeatures(object = so_diab)) -->
<!-- ElbowPlot(so_diab) -->
<!-- # Cluster cells -->
<!-- so_diab <- FindNeighbors(so_diab) -->
<!-- so_diab <- FindClusters(so_diab) -->
<!-- # Perform UMAP and tSNE -->
<!-- # so_diab@reductions$umap <- NULL -->
<!-- # so_diab@meta.data$umap_ <- NULL -->
<!-- so_diab <- RunUMAP(so_diab, dims = 1:15) -->

<!-- # Plot UMAP and label by cell type stored in KPMP_celltype -->
<!-- DimPlot(so_diab, reduction = "umap.rpca", group.by = "medication",raster=F) + -->
<!--   ggtitle("UMAP Among Type 2 Diabetes by Medication Status") + -->
<!--   labs(x = "umap_1", y = "umap_2")  -->

<!-- # Using geom_label instead of geom_text for labels with a box -->
<!-- DimPlot(so_diab, reduction = "umap", group.by = "medication", raster = FALSE) + -->
<!--   geom_label(aes(label = KPMP_celltype), color = "black", size = 3, fill = "white", label.padding = 0.2) + -->
<!--   ggtitle("UMAP by Medication and Labeled by KPMP Cell Type") + -->
<!--   labs(x = "UMAP 1", y = "UMAP 2") -->


<!-- #Type 2 diabetes + OB -->
<!-- so_diab <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_diab <- NormalizeData(so_diab) -->
<!-- so_diab <- ScaleData(so_diab) -->
<!-- so_diab <- RunPCA(so_diab, features = VariableFeatures(object = so_diab)) -->
<!-- ElbowPlot(so_diab) -->
<!-- # Cluster cells -->
<!-- so_diab <- FindNeighbors(so_diab) -->
<!-- so_diab <- FindClusters(so_diab) -->
<!-- # Perform UMAP and tSNE -->
<!-- # so_diab@reductions$umap <- NULL -->
<!-- # so_diab@meta.data$umap_ <- NULL -->
<!-- so_diab <- RunUMAP(so_diab, dims = 1:15) -->

<!-- # Plot UMAP and label by cell type stored in KPMP_celltype -->
<!-- DimPlot(so_diab, reduction = "umap.rpca", group.by = "medication",raster=F) + -->
<!--   ggtitle("UMAP Among T2D + OB by Medication Status") + -->
<!--   labs(x = "umap_1", y = "umap_2")  -->

<!-- # Using geom_label instead of geom_text for labels with a box -->
<!-- DimPlot(so_diab, reduction = "umap", group.by = "medication", raster = FALSE) + -->
<!--   geom_label(aes(label = KPMP_celltype), color = "black", size = 3, fill = "white", label.padding = 0.2) + -->
<!--   ggtitle("UMAP by Medication and Labeled by KPMP Cell Type") + -->
<!--   labs(x = "UMAP 1", y = "UMAP 2") -->
<!-- ``` -->


# 3. Outline
### a. Comparison Groups of Interest
1. T2D vs. OB
2. T2D vs. T1D
3. T2D vs. LC
4. T2D/OB vs. LC
5. T2D/OB vs. T1D

### b. Disease Categories of Interest
1. Albuminuria (Y/N) - UACR >= 30 mg/g
a. T2D
b. T2D/OB
2. UACR >= 100mg/g
3. UACR >= 300mg/g

### c. Medication Groups of Interest
1. SLGT2i+/GLP-1ra+
2. SGLT2i-/GLP-1ra+
3. SGLT2i+/GLP-1ra-
4. SGLT2i-/GLP-1ra-

### d. Additional Medications of Interest
Metformin, Insulin, ACEi/ARB

### e. Characteristics of Interest
UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV

### f. Gene Sets of Interest
1. All Genes
2. Senescence
3. Metabolism
4. Inflammation

### g. Analysis Plan
1. All Genes
a. DEGS + GSEA + IPA
b. LMEM + IPA
c. Pseuodbulk all cell types
d. Cell Specific (PT, TAL, EC, etc.)
2. Pathway Gene Sets
a. DEGS
b. LMEM
c. Pseudobulk all cell types
d. Cell Specific (PT, TAL, EC, etc. )

# 3. Descriptive Statistics & Visualizations
```{r, echo = F,warning=F}
invisible(gc())

#Load in med data to update medication information
med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))
med <- med %>%
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
meta_kidney_sc <-  so_kidney_sc@meta.data
med <- med %>%
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain
med <- med %>%
  filter(record_id %in% meta_kidney_sc$record_id) #81 remain
med$mrn <- as.numeric(med$mrn)
rm(meta_kidney_sc)

# saveRDS(so_kidney_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds"))

#Load metadata for descriptive stats
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv"))
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata2.csv")) #all metadata vars
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))
harm_meta_data <- harm_meta_data %>%
  dplyr::select(-X)
harm_meta_data <- harm_meta_data %>%
  filter(kit_id %in% unique(so_kidney_sc$kit_id))
harm_meta_data <- tidylog::full_join(med,harm_meta_data,by=c("mrn"))
rm(med)

#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>%
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>%
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>%
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
#mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups:
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
harm_meta_data <- harm_meta_data %>%
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","group","group2")))

# #Number of Individuals Per Study
# harm_summary1 <- harm_meta_data %>%
#   group_by(study) %>%
#   summarize(n=n())
#
# #Number of Individuals Per Disease Group
# harm_summary2 <- harm_meta_data %>%
#   group_by(group) %>%
#   summarize(n=n())
#
# #Number of Individuals Per Medication Group
# harm_summary3 <- harm_meta_data %>%
#   group_by(medication) %>%
#   summarize(n=n())
#
# #Number of Individuals per med group & disease group
# harm_summary4 <- harm_meta_data %>%
#   group_by(group,medication) %>%
#   summarize(n=n())

#Note: 83 participants have sc and metadata
#Covars of interest:
#UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV
dat <- harm_meta_data
rm(harm_meta_data)
label(dat$age)      <- "Age (y)"
label(dat$sex)      <- "Sex"
label(dat$hba1c) <- "HbA1c (%)"
dat$medication <- case_when(dat$medication=="no_med"~"No Medication",
                            dat$medication=="sglt2"~"Exclusive SLGT2i",
                            dat$medication=="glp1"~"Exclusive GLP-1ra",
                            dat$medication=="glp1_sglt2"~"GLP-1ra + SGLT2i")
#dat$medication <- factor(dat$medication)
label(dat$medication) <- "Medication Status"
label(dat$ace_inhibitor) <- "ACE Inhibitor (Yes vs. No)"
label(dat$angiotensin_receptor_blocker) <- "Angiotensin Receptor Blocker (Yes vs. No)"
label(dat$raasi_1) <- "ACEi/ARB (Yes vs. No)"
label(dat$sbp) <- "Systolic Blood Pressure (mmHg)"
label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$map) <- "Mean Arterial Pressure (mmHg)"
label(dat$erpf_raw_plasma) <- "Estimated Renal Plasma Flow (mL/min)"
label(dat$bmi) <- "Body Mass Index (kg/m)"
label(dat$dexa_fat_kg) <- "DEXA Fat Mass (kg)"
label(dat$dexa_lean_kg) <- "DEXA Lean Mass (kg)"
label(dat$total_kidney_volume_ml) <- "Total Kidney Volume (mL)"
label(dat$ht_adj_tkv) <- "Height-Adjusted Total Kidney Volume (mL/m)"
label(dat$gfr_raw_plasma) <- "Glomerular Filtration Rate (mL/min)"
label(dat$eGFR_CKD_epi) <- "Estimated Glomerular Filtration Rate (mL/min/1.73 m)"
label(dat$race_ethnicity)    <- "Race/Ethnicity"
label(dat$triglycerides) <-  "Triglycerides (mg/dL)"
label(dat$group) <- "Disease Status"
label(dat$group2) <- "Disease Status"
label(dat$metformin_timepoint) <- "Metformin (Yes vs. No)"
label(dat$mfm_1) <- "Metformin (Yes vs. No)"
label(dat$insulin_med_timepoint) <- NULL
label(dat$insulin_injections_timepoint) <- NULL
label(dat$insulin_1) <- NULL

#race_ethnicity
## Table 1. Descriptive Statistics by Study  ----
table1(~ age + sex + bmi + map + group + medication | study, data=dat,overall=F)
# t1 <- table1(~ age + sex + bmi + map + group + medication | study, data=dat,overall=F)
# kable(t1,caption="Table 1. Descriptive Statistics by Study")
# table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
#          sbp + dbp + map +
#          triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
#          erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
#          ace_inhibitor + angiotensin_receptor_blocker+
#          + group + medication | study, data=dat, overall=F)
# table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
#          sbp + dbp + map +
#          triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
#          erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
#          ace_inhibitor + angiotensin_receptor_blocker+
#          + group + medication | study, data=dat, overall=F)
## Table 2. Table 2. Descriptive Statistics by Disease Status ----
table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
         sbp + dbp + map +
         triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
         erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
         raasi_1+mfm_1+insulin_1+medication | group2, data=dat)
#Med table
# table1(~insulin_injections_timepoint + insulin_med_timepoint+insulin_1| group, data=dat)
# kable(t2,caption="Table 2. Descriptive Statistics by Disease Status")

#Table 3. Table 3. Descriptive Statistics by Disease Status (T2D & OB Combined) ----
table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
         sbp + dbp + map +
         triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
         erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
         ace_inhibitor + angiotensin_receptor_blocker+metformin_timepoint+insulin_med_timepoint+
         + medication | group2, data=dat)
# kable(t3,caption="Table 3. Descriptive Statistics by Disease Status (T2D & OB Combined)")

#Compare meds and groups
# table1(~ medication + insulin_injections_timepoint +insulin_med_timepoint+ ace_inhibitor+angiotensin_receptor_blocker |group,data=dat,overall=F)
# table1(~ medication + insulin_injections_timepoint+insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker  |group2,data=dat,overall=F)
# table1(~ medication +insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker + metformin_timepoint  |group2,data=dat,overall=F)
# table1(~ medication + insulin_injections_timepoint+insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker + metformin_timepoint  |group,data=dat,overall=F)
rm(dat)

#Visualize UMAP


```

# 4. Pseudobulk Analysis
## A. All Genes
### i. DEGS + IPA
#### a. Pseuodbulk all cell types

```{r}
#Average
#Select full gene set
# all_genes <- rownames(so_kidney_sc) #9276 genes for 81 people
all_genes <- rownames(so_kidney_sc) #9276 genes for 81 people
genes <- colnames(data_pb)[which(colnames(data_pb) %in% all_genes)]

#PT
#Type 2 Diabetes
#SGLT1 + GLP1 vs. GLP1
data_pb1 <- data_pb %>% 
  filter(celltype=="PT") %>% 
  filter(group=="Type 2 Diabetes") %>% 
  filter(medication=="glp1_sglt2" | medication=="glp1") %>% 
  mutate(across(all_of(genes),~log2(. + 1)))
title <- paste0("Differential Expression in SGLT2+GLP1 vs. GLP1 in Type 2 Diabetes in PT Cells")

total_results <- data.frame()
for (gene in genes) {
# Perform a t-test
t_test_result <- t.test(data_pb1[[gene]] ~ data_pb1$medication)
pvalue <- t_test_result$p.value
log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
mean_exp <- t_test_result$estimate[[2]]
mean_ref <- t_test_result$estimate[[1]]
results <- data.frame(Gene=gene,Log2FC=log2fc,Mean_Exp=mean_exp,Mean_Ref=mean_ref,Pvalue=pvalue)
total_results <- rbind(total_results,results)
}
total_results$fdr <- p.adjust(total_results$Pvalue,method="fdr")
# Sort the data to get the top 10 significant positive and negative genes
top_positive <- total_results %>%
  filter(Log2FC > 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

top_negative <- total_results %>%
  filter(Log2FC < 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

# Combine top positive and negative genes for labeling
top_genes <- rbind(top_positive, top_negative)

# Create the volcano plot
ggplot(total_results, aes(x = Log2FC, y = -log10(Pvalue))) +
  geom_point(aes(color = ifelse(Log2FC > 0 & fdr < 0.05, "red", 
                                ifelse(Log2FC < 0 & fdr < 0.05, "blue", "gray"))),
             size = 2, alpha = 0.7) +  # Set a constant size for all points
  scale_color_identity() +  # Use the exact colors specified in the color aesthetic
  labs(title = title, x = "Log2 Fold Change", y = "-log10(P-value)") +
  theme_minimal() +
  theme(legend.position = "none") +  # Hide legend
  # geom_text(data = top_genes, aes(label = Gene), vjust = 1, hjust = 1, size = 3, color = "black")+
   geom_text_repel(data = top_genes, aes(label = Gene), 
                  size = 3, color = "black", box.padding = 0.5, point.padding = 0.5, 
                  segment.color = "grey50", max.overlaps = 10) 
write.csv(total_results,fs::path(dir.ipa,"PT_T2D_SLGT2_GLP1_vs_GLP1.csv"))

# #SGLT1 + GLP1 vs. GLP1
# data_pb1 <- data_pb_sum %>% 
#   filter(celltype=="PT") %>% 
#   filter(group=="Type 2 Diabetes") %>% 
#   filter(medication=="glp1_sglt2" | medication=="glp1") %>% 
#   mutate(across(all_of(genes),~log2(. + 1)))
# title <- paste0("Differential Expression in SGLT2+GLP1 vs. GLP1 in Type 2 Diabetes in PT Cells")
# 
# total_results <- data.frame()
# for (gene in genes) {
# # Perform a t-test
# t_test_result <- t.test(data_pb1[[gene]] ~ data_pb1$medication)
# pvalue <- t_test_result$p.value
# log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
# mean_exp <- t_test_result$estimate[[2]]
# mean_ref <- t_test_result$estimate[[1]]
# results <- data.frame(Gene=gene,Log2FC=log2fc,Mean_Exp=mean_exp,Mean_Ref=mean_ref,Pvalue=pvalue)
# total_results <- rbind(total_results,results)
# }
# total_results$fdr <- p.adjust(total_results$Pvalue,method="fdr")
# 
# # Create the volcano plot
# ggplot(total_results, aes(x = Log2FC, y = -log10(Pvalue))) +
# geom_point(aes(color = ifelse(Log2FC > 0 & Pvalue < 0.05, "red", 
#                                 ifelse(Log2FC < 0 & Pvalue < 0.05, "blue", "gray"))),
#              size = 2, alpha = 0.7)+
#   scale_color_identity() +  # Use the exact colors specified in the color aesthetic
#   labs(title = title, x = "Log2 Fold Change", y = "-log10(P-value)") +
#   theme_minimal() +
#   theme(legend.position = "none")  # Hide legend
# write.csv(total_results,fs::path(dir.ipa,"PT_T2D_SLGT2_GLP1_vs_GLP1_sum.csv"))

#SGLT1 + GLP1 vs. SGLT2
data_pb1 <- data_pb %>% 
  filter(celltype=="PT") %>% 
  filter(group=="Type 2 Diabetes") %>% 
  filter(medication=="glp1_sglt2" | medication=="sglt2") %>% 
  mutate(across(all_of(genes),~log2(. + 1)))
title <- paste0("Differential Expression in SGLT2+GLP1 vs. SGLT2 in Type 2 Diabetes in PT Cells")

total_results <- data.frame()
for (gene in genes) {
# Perform a t-test
t_test_result <- t.test(data_pb1[[gene]] ~ data_pb1$medication)
pvalue <- t_test_result$p.value
log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
results <- data.frame(Gene=gene,Log2FC=log2fc,Pvalue=pvalue)
total_results <- rbind(total_results,results)
}
total_results$fdr <- p.adjust(total_results$Pvalue,method="fdr")
top_positive <- total_results %>%
  filter(Log2FC > 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

top_negative <- total_results %>%
  filter(Log2FC < 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

# Combine top positive and negative genes for labeling
top_genes <- rbind(top_positive, top_negative)

# Create the volcano plot
ggplot(total_results, aes(x = Log2FC, y = -log10(Pvalue))) +
  geom_point(aes(color = ifelse(Log2FC > 0 & fdr < 0.05, "red", 
                                ifelse(Log2FC < 0 & fdr < 0.05, "blue", "gray"))),
             size = 2, alpha = 0.7) +  # Set a constant size for all points
  scale_color_identity() +  # Use the exact colors specified in the color aesthetic
  labs(title = title, x = "Log2 Fold Change", y = "-log10(P-value)") +
  theme_minimal() +
  theme(legend.position = "none") +  # Hide legend
  # geom_text(data = top_genes, aes(label = Gene), vjust = 1, hjust = 1, size = 3, color = "black")+
   geom_text_repel(data = top_genes, aes(label = Gene), 
                  size = 3, color = "black", box.padding = 0.5, point.padding = 0.5, 
                  segment.color = "grey50", max.overlaps = 10) 
write.csv(total_results,fs::path(dir.ipa,"PT_T2D_SLGT2_GLP1_vs_SGLT2.csv"))

#SGLT1 + GLP1 vs. SGLT2
data_pb1 <- data_pb %>% 
  filter(celltype=="PT") %>% 
  filter(group=="Type 2 Diabetes") %>% 
  filter(medication=="glp1_sglt2" | medication=="sglt2") %>% 
  mutate(across(all_of(genes),~log2(. + 1)))
title <- paste0("Differential Expression in SGLT2+GLP1 vs. SGLT2 in Type 2 Diabetes in PT Cells")

total_results <- data.frame()
for (gene in genes) {
# Perform a t-test
t_test_result <- t.test(data_pb1[[gene]] ~ data_pb1$medication)
pvalue <- t_test_result$p.value
log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
results <- data.frame(Gene=gene,Log2FC=log2fc,Pvalue=pvalue)
total_results <- rbind(total_results,results)
}
total_results$fdr <- p.adjust(total_results$Pvalue,method="fdr")
top_positive <- total_results %>%
  filter(Log2FC > 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

top_negative <- total_results %>%
  filter(Log2FC < 0 & fdr < 0.05) %>%
  arrange(fdr) %>%
  head(10)

# Combine top positive and negative genes for labeling
top_genes <- rbind(top_positive, top_negative)

# Create the volcano plot
ggplot(total_results, aes(x = Log2FC, y = -log10(Pvalue))) +
  geom_point(aes(color = ifelse(Log2FC > 0 & fdr < 0.05, "red", 
                                ifelse(Log2FC < 0 & fdr < 0.05, "blue", "gray"))),
             size = 2, alpha = 0.7) +  # Set a constant size for all points
  scale_color_identity() +  # Use the exact colors specified in the color aesthetic
  labs(title = title, x = "Log2 Fold Change", y = "-log10(P-value)") +
  theme_minimal() +
  theme(legend.position = "none") +  # Hide legend
  # geom_text(data = top_genes, aes(label = Gene), vjust = 1, hjust = 1, size = 3, color = "black")+
   geom_text_repel(data = top_genes, aes(label = Gene), 
                  size = 3, color = "black", box.padding = 0.5, point.padding = 0.5, 
                  segment.color = "grey50", max.overlaps = 10) 
write.csv(total_results,fs::path(dir.ipa,"PT_T2D_SLGT2_GLP1_vs_SGLT2.csv"))
```

```{r,echo=F,warning=F,fig.width=15, fig.height=9}

#Average
#Select full gene set
# all_genes <- rownames(so_kidney_sc) #9276 genes for 81 people
all_genes <- rownames(so_kidney_sc) #9276 genes for 81 people

#PT
# so_pb <- subset(so_kidney_sc,celltype=="PT")
genes <- colnames(data_pb)[which(colnames(data_pb) %in% all_genes)]
data_pb1 <- data_pb %>% 
  filter(celltype=="PT") %>% 
  filter(group=="Type 2 Diabetes") %>% 
  filter(medication=="glp1_sglt2" | medication=="glp1") %>% 
  mutate(across(all_of(genes),~log2(. + 1)))
# data_pb_sum1 <- data_pb_sum %>% 
#   filter(celltype=="PT") %>% 
#   filter(group=="Type 2 Diabetes") %>% 
#   filter(medication=="glp1_sglt2" | medication=="glp1")%>% 
#   mutate(across(all_of(genes),~log2(. + 1)))

# so_pb <- so_kidney_sc
#Medications (AMong Type 2s)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
# so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32

#Combo vs. GLP1
# so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
# so_subgroup$medication <- factor(so_subgroup$medication)

#Log transform genes
#Mean in group1 and mean in group2 

#ttest of log2(group1),log2(group2)
#Difference of means is the effect size
#Pathways with fold change from mean diff and pvalue from ttest
total_results <- data.frame()
for (gene in genes) {
# Perform a t-test
t_test_result <- t.test(data_pb1[[gene]] ~ data_pb1$medication)
pvalue <- t_test_result$p.value
log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
results <- data.frame(Gene=gene,Log2FC=log2fc,Pvalue=pvalue)
total_results <- rbind(total_results,results)
}
total_results$fdr <- p.adjust(total_results$Pvalue,method="fdr")

# Create the volcano plot
ggplot(total_results, aes(x = Log2FC, y = -log10(Pvalue))) +
  geom_point(aes(color = Pvalue < 0.05, size = abs(Log2FC)), alpha = 0.7) + # Color points by significance, size by Log2FC
  scale_color_manual(values = c("gray", "red")) +  # Color significance points in red
  labs(title = "Means", x = "Log2 Fold Change", y = "-log10(P-value)") +
  theme_minimal() +
  theme(legend.position = "none")  # Hide legend

# #Pathways with fold change from mean diff and pvalue from ttest
# total_results2 <- data.frame()
# for (gene in genes) {
# # Perform a t-test
# t_test_result <- t.test(data_pb_sum1[[gene]] ~ data_pb_sum1$medication)
# pvalue <- t_test_result$p.value
# log2fc <- t_test_result$estimate[[2]]-t_test_result$estimate[[1]]
# results2 <- data.frame(Gene=gene,Log2FC=log2fc,Pvalue=pvalue)
# total_results2 <- rbind(total_results2,results2)
# }
# total_results2$fdr <- p.adjust(total_results2$Pvalue,method="fdr")
# 
# # Create the volcano plot
# ggplot(total_results2, aes(x = Log2FC, y = -log10(Pvalue))) +
#   geom_point(aes(color = Pvalue < 0.05, size = abs(Log2FC)), alpha = 0.7) + # Color points by significance, size by Log2FC
#   scale_color_manual(values = c("gray", "red")) +  # Color significance points in red
#   labs(title = "Sum", x = "Log2 Fold Change", y = "-log10(P-value)") +
#   theme_minimal() +
#   theme(legend.position = "none")  # Hide legend
# 
# 

# all_genes <- rownames(so_subgroup)
# Idents(so_subgroup) <- so_subgroup$celltype
# degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT")

#Summed
all_genes <- rownames(pseudobulked_data_sum) #9276 genes for 81 people

#PT
# so_pb <- subset(so_kidney_sc,celltype=="PT")
so_pb <- subset(pseudobulked_data_sum,celltype=="PT")

# so_pb <- so_kidney_sc
#Medications (AMong Type 2s)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
# Idents(so_subgroup) <- so_subgroup$celltype
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT")




########
#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT")

#TAL
so_pb <- subset(so_kidney_sc,celltype=="TAL")
# so_pb <- so_kidney_sc
#Medications (AMong Type 2s)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="TAL")


#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="TAL")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="TAL")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="TAL")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="TAL")

#EC
so_pb <- subset(so_kidney_sc,celltype=="EC")
# so_pb <- so_kidney_sc
#Medications (AMong Type 2s)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC")


#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC")


#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC")


```

#### b. Specific Pathway Genes
##### i. PT Cells
```{r, echo=F, warning=F, fig.width=10,fig,height=10}

#Table for pvalues and dot plot for expression


#PT Cells
#Hemodynamics
hemo1 <- c("SLC9A3", "SLC12A1", "SLC12A3", "SCNN1G", "SLC5A2", "SLC5A1", 
           "SLC22A6", "SLC22A8", "SLC22A2", "SLC22A11", "SLC22A12", 
           "SLC34A1", "SLC34A3", "SLC13A3", "SLC13A2", "SLC16A9", "SLC4A4")
#Check genes are in seurat object
hemo1 <- hemo1[which(hemo1 %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo1))
degs_fxn3(so=so_subgroup2,cell=NULL,gene_set=hemo1,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Hemodynamics")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo1))
degs_fxn3(so=so_subgroup,cell=NULL,gene_set=hemo1,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Hemodynamics")

# m = FindMarkers(so_subgroup, features = hemo1,group.by = "medication",ident.1 = "sglt2",
#                   ident.2 = "glp1_sglt2", subset.ident = "PT", verbose = F)
# de.markers(so_subgroup, hemo1, "medication", id2 = "sglt2", id1 = "glp1_sglt2", "PT", "PT")
# 
# de.markers(subset(so, KPMP_celltype == "aPT"), genes, "Group", id2 = "HC", id1 = "T1D", NULL, ".aPT")
# 
# dp.formatted <- function(seurat_object, genes, celltype, group.by, m,
#                            colorlow = "#83c5be", colormid = "#f4f1bb", colorhigh = "#d90429")
#   

# 
# pt.plot_table <- dp.formatted(seurat_object = so, genes = genes_subset, group.by = "Group",
#                    celltype = "PT", m = m)
#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo1))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=hemo1,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Hemodynamics")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo1))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=hemo1,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Hemodynamics")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo1))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=hemo1,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Hemodynamics")

###########
infl <- c("IL6", "IL8", "CXCL1", "CXCL2", "CCL20", "ICAM1", "TNFAIP3", "NAF1", "TRAF1", "TNF")
#Check genes are in seurat object
infl<- infl[which(infl %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=infl,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Inflammation1")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Inflammation1")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Inflammation1")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Inflammation1")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Inflammation1")
#############
infl2 <-  c("C3", "C4", "CFB", "C5", "C6", "C7", "C8", "C9", "BMP2", "PTPRC", "CCL2")
#Check genes are in seurat object
infl2<- infl2[which(infl2 %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl2))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=infl2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="infl2ammation2")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="infl2ammation2")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="infl2ammation2")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="infl2ammation2")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), infl2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=infl2,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="infl2ammation2")
##############
ros <- c("SOD1", "SOD2", "NOX4", "DAO", "NRF2", "FOXO1", "FOXO3", "FOXO4", "HMOX1")
#Check genes are in seurat object
ros<- ros[which(ros %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=ros,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species")
################
ros2 <- c("ALOX12", "ALOX15", "ALOX5", "ALOX5AP")
#Check genes are in seurat object
ros2<- ros2[which(ros2 %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species2")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species2")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species2")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species2")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=ros2,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Reactive_Oxygen_Species2")
########
injury <- c("LRP2", "PPARA", "PPARGC1A", "SLC22A1", "SLC12A3")
#Check genes are in seurat object
injury<- injury[which(injury %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=injury,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers")

#############
injury2 <- c("CST3", "HAVCR1", "LCN2")
#Check genes are in seurat object
injury2<- injury2[which(injury2 %in% rownames(so_kidney_sc))]

#PT
so_pb <- subset(so_kidney_sc,celltype=="PT")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury2))
#Run
degs_fxn3(so=so_subgroup2,cell="PT",gene_set=injury2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers22")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers22")

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers22")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers22")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), injury2))
degs_fxn3(so=so_subgroup,cell="PT",gene_set=injury2,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="PT",pathway="Injury_Markers22")
```

##### ii. TAL Cells
```{r, echo=F,warning=F,fig.width=5,fig.height=8}
hemo2 <- c("SLC12A1", "SLC22A3", "SLC22A1", "SLC4A1", "SLC19A1", 
           "SLC12A3", "KCNJ1", "CLCNKB", "CASR", "SLC8A1", # Transporters and Ion Channels
           "SLC26A3", "SLC9A3", "SLC9A1", "SLC5A1", "SLC5A2", # Additional transporters
           "ATP2B1", "ATP2B2", "ATP1A1", "ATP1B1", "KCNJ10",  # Ion pumps and channels
           "PTH1R", "PTH2R", "RFX6", "CLCNKA", "TMEM16A",     # Hormone receptors, chloride channels
           "SLC13A2", "SLC13A5", "SLC16A9", "SLC36A4", "SLC35F2", # Further solute carriers
           "NKCC1", "SPHK1", "OCTN2", "HMGCR")                # Other transporters and enzymes

#Check genes are in seurat object
hemo2 <- hemo2[which(hemo2 %in% rownames(so_kidney_sc))]

#TAL
so_pb <- subset(so_kidney_sc,celltype=="TAL")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo2))
#Run
degs_fxn3(so=so_subgroup2,cell="TAL",gene_set=hemo2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="TAL",pathway="Hemodynamics")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo2))
degs_fxn3(so=so_subgroup,cell="TAL",gene_set=hemo2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="TAL",pathway="Hemodynamics")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo2))
degs_fxn3(so=so_subgroup,cell="TAL",gene_set=hemo2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="TAL",pathway="Hemodynamics")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo2))
degs_fxn3(so=so_subgroup,cell="TAL",gene_set=hemo2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="TAL",pathway="Hemodynamics")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo2))
degs_fxn3(so=so_subgroup,cell="TAL",gene_set=hemo2,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="TAL",pathway="Hemodynamics")

```

##### iii. EC Cells
```{r, echo=F,warning=F,fig.width=5,fig.height=8}
hemo3 <- c("RYR2", "RYR3", "KCNMA1", "KCNN1", "KCNN2", "KCNN3", 
           "KCNJ2", "KCNJ3", "KCNJ5", "KCNJ10", "KCNK3", 
           "KCNK9", "KCNK10", "SLC22A5", "ATP1A1", "ATP1B1", 
           "PIK3CA", "PIK3CB", "NOS3", "EDN1", "VEGFA", 
           "CAV1", "TMEM16A", "CAV2", "KLF2", "KLF4")

#Check genes are in seurat object
hemo3 <- hemo3[which(hemo3 %in% rownames(so_kidney_sc))]

#EC
so_pb <- subset(so_kidney_sc,celltype=="EC")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo3))
#Run
degs_fxn3(so=so_subgroup2,cell="EC",gene_set=hemo3,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics K+ Channels")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo3))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo3,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics K+ Channels")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo3))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo3,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics K+ Channels")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo3))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo3,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics K+ Channels")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo3))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo3,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics K+ Channels")


##########
hemo4 <- c("NOS3", "NOS1", "NOS2", "COX1", "COX2", 
           "PTGS1", "PTGS2", "PTGER1", "PTGER2", "PTGER3", 
           "PTGER4", "PTGES3", "PTGDS", "ALOX5", "ALOX12", 
           "ALOX15", "PTGF", "CYP450", "EDNRA", "EDNRB", 
           "VEGFA", "EDN1", "VEGFR2", "KLF2", "KLF4", 
           "SERPINE1", "MMP9", "MMP2", "ADAMTS1", "TGFBR1", 
           "TGFBR2")
#Check genes are in seurat object
hemo4 <- hemo4[which(hemo4 %in% rownames(so_kidney_sc))]

#EC
so_pb <- subset(so_kidney_sc,celltype=="EC")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo4))
#Run
degs_fxn3(so=so_subgroup2,cell="EC",gene_set=hemo4,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Signaling")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo4))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo4,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Signaling")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo4))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo4,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Signaling")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo4))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo4,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Signaling")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo4))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo4,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Signaling")

#########
# vascular_remodeling_endothelial
hemo5 <- c("ICAM1","VCAM1","NOS3","CD31","VWF","KDR","PLVAP","VEGFA")
#Check genes are in seurat object
hemo5 <- hemo5[which(hemo5 %in% rownames(so_kidney_sc))]

#EC
so_pb <- subset(so_kidney_sc,celltype=="EC")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo5))
#Run
degs_fxn3(so=so_subgroup2,cell="EC",gene_set=hemo5,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Vascular Remodeling")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo5))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo5,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Vascular Remodeling")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo5))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo5,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Vascular Remodeling")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo5))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo5,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Vascular Remodeling")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), hemo5))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=hemo5,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="Hemodynamics Vascular Remodeling")

#########
#ros_regulation_endothelial
ros2 <- c("SOD1","SOD2","NOX2","ERO1","ACOX","DAO","XO","HMOX1","KEAP1","NRF2","FOXO1","FOXO3","FOXO4")
#Check genes are in seurat object
ros2 <- ros2[which(ros2 %in% rownames(so_kidney_sc))]

#EC
so_pb <- subset(so_kidney_sc,celltype=="EC")
so_diab <- subset(so_pb,group=="Type 2 Diabetes") #32
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
#Run
degs_fxn3(so=so_subgroup2,cell="EC",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")
degs_fxn3(so=so_subgroup2,cell="EC",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
length(unique(so_subgroup$kit_id)) #12
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=ros2,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")

#Medications (AMong T2D + OB)
#Filter to Type_2_Diabetes only since no other participants are on these main meds
so_diab <- subset(so_pb,group2=="T2D + OB")
length(unique(so_diab$kit_id)) #35
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
so_subgroup2 <- subset(so_subgroup, features = intersect(rownames(so_subgroup), ros2))
degs_fxn3(so=so_subgroup,cell="EC",gene_set=ros2,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30,cellname="EC",pathway="ROS Regulation")

```

<!-- # 5. Non-Pseudobulk Analysis -->
<!-- ## A. All Genes -->
<!-- ### i. DEGS + IPA -->
<!-- #### a. Pseuodbulk all cell types -->
<!-- ```{r,echo=F,warning=F,fig.width=15, fig.height=9} -->
<!-- #Select full gene set -->
<!-- all_genes <- rownames(so_kidney_sc) -->

<!-- #Disease Groups -->
<!-- #1. T2D vs. OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30) -->
<!-- # ipa <- read_delim() -->

<!-- #2. T2D vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->


<!-- #3. T2D vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->


<!-- #4. T2D/OB vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="Yes",top_gsea = 30) -->

<!-- #5. T2D/OB vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="Yes",top_gsea = 30) -->

<!-- #1. Albuminuria (Y/N) - UACR >= 30 mg/g -->
<!-- so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->

<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->


<!-- #2. UACR >= 100mg/g -->
<!-- so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->

<!-- #"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->

<!-- #3. UACR >= 300mg/g -->
<!-- so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->

<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->

<!-- #Medications (AMong Type 2s) -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->


<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->


<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->


<!-- #Medications (AMong T2D + OB) -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->


<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->


<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- ##RAASI - Type 2 Diabetes -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!-- print (result) -->

<!-- ##RAASI - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!-- print (result) -->

<!-- #Metformin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!-- print (result) -->
<!-- # -->
<!-- #Metformin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!-- print (result) -->

<!-- #Insulin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->

<!-- # -->
<!-- #Insulin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!-- print (result) -->
<!-- ``` -->

<!-- #### b. PT & TAL Pseudobulk -->
<!-- ```{r,echo=F,warning=F,fig.width=15, fig.height=9} -->
<!-- #Pseuodbulk Tall and PT cells -->
<!-- so_kidney_sc$cell_type2 <- ifelse(grepl("PT-",so_kidney_sc$celltype_rpca),"PT", -->
<!--                                   ifelse(grepl("TAL-",so_kidney_sc$celltype_rpca),"TAL",NA)) -->
<!-- so_kidney_sc$cell_type2 <- as.character(so_kidney_sc$cell_type2) -->
<!-- all_cell_types <- unique(so_kidney_sc$cell_type2) -->
<!-- all_cell_types <- all_cell_types[-1] -->
<!-- Idents(so_kidney_sc) <- so_kidney_sc$cell_type2 -->

<!-- #1. T2D vs. OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- #Diabetes yes vs. no -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. T2D vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. T2D vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #4. T2D/OB vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group2=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #5. T2D/OB vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group2=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #1. Albuminuria (Y/N) - UACR >= 30 mg/g -->
<!-- so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. UACR >= 100mg/g -->
<!-- so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #3. UACR >= 300mg/g -->
<!-- so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Medications (AMong Type 2s) -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->

<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #T2D + OB -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Other med groups -->
<!-- #RAASI - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->
<!-- # -->
<!-- ##RAASI - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- ``` -->



<!-- #### c. KPMP Cell Type 1 (PT, TAL, EC, etc.) -->
<!-- ```{r,echo=F,warning=F,fig.width=15, fig.height=9} -->

<!-- so_kidney_sc$celltype_rpca <- as.character(so_kidney_sc$celltype_rpca) -->
<!-- all_cell_types <- unique(so_kidney_sc$celltype_rpca) -->
<!-- Idents(so_kidney_sc) <- so_kidney_sc$celltype_rpca -->
<!-- all_cell_types <- all_cell_types[-29] -->
<!-- #Select full gene set -->
<!-- all_genes <- rownames(so_kidney_sc) -->

<!-- #1. T2D vs. OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- #Diabetes yes vs. no -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. T2D vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #3. T2D vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #4. T2D/OB vs. LC -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group2=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL, exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #5. T2D/OB vs. T1D -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group2=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #1. Albuminuria (Y/N) - UACR >= 30 mg/g -->
<!-- so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. UACR >= 100mg/g -->
<!-- so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- # all_cell_types2 <- all_cell_types[-29] -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. UACR >= 300mg/g -->
<!-- so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Medications (Among Type 2 Diabetes) -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->

<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Medications (Among T2D + OB) -->
<!-- #Filter to T2D + OB only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->

<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Other med groups -->
<!-- #RAASI - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- ##RAASI - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - T2D + OB -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kidney_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(26,32)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- ``` -->

<!-- #### d. KPMP Cell Type 2 (PT, TAL, EC, etc.) -->
<!-- ```{r,echo=F,warning=F,fig.width=15, fig.height=9} -->
<!-- so_kpmp_sc$KPMP_celltype <- as.character(so_kpmp_sc$KPMP_celltype) -->
<!-- all_cell_types <- unique(so_kpmp_sc$KPMP_celltype) -->
<!-- Idents(so_kpmp_sc) <- so_kpmp_sc$KPMP_celltype -->
<!-- all_cell_types <- all_cell_types[-40] -->
<!-- #Select full gene set -->
<!-- all_genes <- rownames(so_kpmp_sc) -->

<!-- #1. T2D vs. OB -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes" | group=="Obese_Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- #Diabetes yes vs. no -->
<!-- for (cell_type in all_cell_types[-42]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type_2_Diabetes",ref_group="Obese_Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. T2D vs. T1D -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes" | group=="Type_1_Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type_2_Diabetes",ref_group="Type_1_Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #3. T2D vs. LC -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes" | group=="Lean_Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-42]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type_2_Diabetes",ref_group="Lean_Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #4. T2D/OB vs. LC -->
<!-- so_subgroup <- subset(so_kpmp_sc,group2=="T2D + OB" | group2=="Lean Control") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL, exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #5. T2D/OB vs. T1D -->
<!-- so_subgroup <- subset(so_kpmp_sc,group2=="T2D + OB" | group2=="Type 1 Diabetes") -->
<!-- so_subgroup$DiseaseGroup <- so_subgroup$group2 -->
<!-- so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #1. Albuminuria (Y/N) - UACR >= 30 mg/g -->
<!-- so_kpmp_sc$albuminuria_30 <- ifelse(so_kpmp_sc$acr_u>=30,"Above_30","Below_30") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. UACR >= 100mg/g -->
<!-- so_kpmp_sc$albuminuria_100 <- ifelse(so_kpmp_sc$acr_u>=100,"Above_100","Below_100") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- # all_cell_types2 <- all_cell_types[-29] -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. UACR >= 300mg/g -->
<!-- so_kpmp_sc$albuminuria_300 <- ifelse(so_kpmp_sc$acr_u>=300,"Above_300","Below_300") -->
<!-- #  a. T2D -->
<!-- so_subgroup <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #  b. T2D/OB -->
<!-- so_subgroup <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Medications (Among Type 2 Diabetes) -->
<!-- #Filter to type 2 diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->

<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Medications (Among T2D + OB) -->
<!-- #Filter to T2D + OB only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_kpmp_sc,group2=="T2D + OB") -->

<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->
<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->
<!--   print(result) -->
<!-- } -->

<!-- #Other med groups -->
<!-- #RAASI - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type_2_Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- ##RAASI - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No") -->
<!-- so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Metformin - T2D + OB -->
<!-- # #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No") -->
<!-- so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - type 2s only -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group=="Type_2_Diabetes") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->

<!-- #Insulin - T2D + OB -->
<!-- #Filter out NAs for ace inhibitors -->
<!-- so_diab <- subset(so_kpmp_sc,group2=="T2D + OB") -->
<!-- so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No") -->
<!-- so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- for (cell_type in all_cell_types[-c(42)]) { -->
<!--   result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30) -->
<!--   print (result) -->
<!-- } -->
<!-- ``` -->

<!-- ### ii. LMEM + IPA -->
<!-- #### a. Pseudobulk All Cell Types -->
<!-- ```{r, echo=F,warning=F,fig.width=15,fig.height=9} -->
<!-- #Characteristics of interest -->
<!-- # UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV -->

<!-- #Write a linear mixed effect model: lmem (Gene Exp = alpha + Beta*Continuous_Var + Covars + individual level random intercept) -->
<!-- # Assuming you have a dataset where: -->
<!-- # - `response` is the dependent variable gene expression (e.g., measurement on cells). -->
<!-- # - `predictor1` and `predictor2` are the fixed effects (e.g., cell-level variables or other covariates). -->
<!-- # - `individual_id` is the ID for each individual. -->
<!-- # - `cell_id` is the ID for each cell nested within an individual. -->

<!-- # Extract the gene expression data for all genes -->
<!-- gene_expression <- as.data.frame(so_kpmp_sc@assays$RNA@layers$data) -->

<!-- # Assign the gene names (row names of Seurat object) as the column names -->
<!-- rownames(gene_expression) <- rownames(so_kpmp_sc) -->
<!-- colnames(gene_expression) <- colnames(so_kpmp_sc) -->
<!-- gene_expression <- t(gene_expression) -->
<!-- gene_expression <- data.frame(gene_expression) -->
<!-- gene_list <- colnames(gene_expression) -->
<!-- #log transform gene expression -->
<!-- gene_expression <- gene_expression %>%  -->
<!--   mutate(across(everything(),~log1p(.))) # log1p is log(x + 1) -->
<!-- gene_expression$cellname <- rownames(gene_expression) -->
<!-- rownames(gene_expression) <- NULL -->

<!-- # Extract the metadata, make sure you select relevant columns (e.g., 'predictor1', 'disease_status') -->
<!-- metadata <- so_kpmp_sc@meta.data -->
<!-- metadata <- metadata %>%  -->
<!--   mutate(across(everything(),~ifelse(.==".",NA,.))) -->
<!-- metadata$cellname <- rownames(metadata) -->
<!-- rownames(metadata) <- NULL -->

<!-- # filter_ids <- unique(data$kit_id[which(data[[x]]==".")]) -->

<!-- # Combine the gene expression data and metadata -->
<!-- # This ensures cells (columns) are aligned properly between the two -->
<!-- data <- tidylog::full_join(metadata,gene_expression,by="cellname") -->
<!-- # unique(data$kit_id) #81 participants -->
<!-- # which(colnames(data) =="cellname") -->
<!-- unique(data$group) -->
<!-- data$group <- factor(data$group,levels = c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes")) -->
<!-- # length(data$hba1c) -->
<!-- # length(which(data$hba1c==".")) #7457 cells missing hba1c -->
<!-- # length(unique(data$kit_id[which(data$hba1c==".")])) #3 people missing HbA1c: "KL-0029407" "KL-0030607" "KL-0031456" -->
<!-- rm(gene_expression,metadata) -->


<!-- #Groups -->
<!-- #1. T2D vs. OB -->
<!-- #Zero inflated negative binomial mixed model/negative binomail mixed model -->
<!-- exp <- "Type_2_Diabetes" -->
<!-- ref <- "Obese_Control" -->
<!-- lmm_fxn(exp = "Type_2_Diabetes",ref = "Obese_Control") -->

<!-- #------------------------------------------------------------------------------------------------------------------------ -->
<!-- #1. Medications among Type 2 Diabetes -->
<!-- #Filter to Type_2_Diabetes only since no other participants are on these main meds -->
<!-- length(unique(data$kit_id)) #81 partiucipants before filtering -->
<!-- data_diab <- data %>%  -->
<!--   filter(group=="Type_2_Diabetes")  #32 participants with Typd 2 -->
<!-- length(unique(data_diab$kit_id)) #32 -->
<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ vs. GLP1 -->
<!-- data_subgroup <- data_diab %>%  -->
<!--   filter(group2=="glp1_sglt2" | group2=="glp1") -->
<!-- length(unique(data_subgroup$kit_id)) #11 -->
<!-- data_subgroup$medication <- factor(data_subgroup$medication) -->
<!-- all_genes <- rownames(data_subgroup) -->
<!-- data_subgroup <- data %>%  -->
<!--   # filter(group=="Type_2_Diabetes" | group=="Obese_Control") -->
<!--   filter(group==ref | group==exp) -->
<!-- data_subgroup$group <- factor(data_subgroup$group) -->
<!-- data_subgroup$group <-relevel(data_subgroup$group, ref = ref) -->
<!-- # full_results <- data.frame() -->
<!-- # Prepare parallel execution of gene models -->
<!-- full_results <- foreach(gene = gene_list[1:48], .combine = rbind, .packages = c("glmmTMB", "dplyr")) %dopar% { -->
<!--   # for (gene in gene_list) { -->
<!--   # for (gene in gene_list[1:20]) { #tester genes -->
<!--   m0 <- as.formula(paste0(gene," ~ group + (1 | kit_id)")) -->
<!--   model <- glmmTMB(m0,data=data_subgroup,family = nbinom2,  # Negative Binomial with a log link -->
<!--                    ziformula = ~ 1) -->
<!--   #If doesnt fit, try nb  -->
<!--   if(summary(model)$coef$zi[1,4]=="NaN") { -->
<!--     model2 <- glmmTMB(m0, data=data_subgroup, -->
<!--                       family = nbinom2) -->
<!--     foldchange <- round(exp(summary(model2)$coef$cond[2,1]) - 1,5) -->
<!--     pval <- summary(model2)$coef$cond[2,4] -->
<!--     model_name <- "Negative Binomial" -->
<!--   }  -->
<!--   if(summary(model)$coef$zi[1,4]!="NaN") { -->
<!--     #Check if zero inflation component is non-significant, if non-sig, run nb -->
<!--     if(summary(model)$coef$zi[1,4]>=0.05) { -->
<!--       model2 <- glmmTMB(m0, data=data_subgroup, -->
<!--                         family = nbinom2) -->
<!--       foldchange <- round(exp(summary(model2)$coef$cond[2,1]) - 1,5) -->
<!--       pval <- summary(model2)$coef$cond[2,4] -->
<!--       model_name <- "Negative Binomial" -->
<!--     }  -->
<!--     #If sig, pull estimates from zinb  -->
<!--     if(summary(model)$coef$zi[1,4]<0.05) { -->
<!--       # foldchange <- round(exp(summary(model)$coef$cond[2,1]) - 1,5) -->
<!--       foldchange <- round(exp(summary(model)$coef$cond[2,1]) - 1,5) -->
<!--       pval <- summary(model)$coef$cond[2,4] -->
<!--       model_name <- "Zero-Inflated Negative Binomial" -->
<!--     } -->
<!--   } -->
<!--   # Collect results for each gene -->
<!--   results <- data.frame(Gene = gene, FoldChange = foldchange, PValue = pval,ModelName=model_name) -->
<!--   return(results) -->
<!--   # results <- data.frame(Gene=gene,FoldChange=foldchange,PValue=pval) -->
<!--   # full_results <- rbind(full_results,results) -->
<!--   # # print(p) -->
<!--   # # return(full_results) -->
<!-- } -->
<!-- #Make volcano plot of all gene results for group -->
<!-- full_results <- full_results %>%  -->
<!--   mutate(fdr=p.adjust(PValue,method="fdr")) -->

<!-- m_top <- full_results -->

<!-- # Transform PValue into -log10(p-value) for the y-axis -->
<!-- m_top$logPValue <- -log10(m_top$PValue) -->

<!-- # Calculate the log2 of the fold change (logFC) -->
<!-- m_top$logFC <- log2(m_top$FoldChange) -->

<!-- # Create a new column for significance based on the p-value cutoff and FC cutoff -->
<!-- m_top$Significance <- ifelse(m_top$fdr<0.05, "Significant", "Not Significant") -->

<!-- # Create a new column to combine both significance and direction (positive or negative) -->
<!-- m_top$Color <- ifelse(m_top$Significance == "Significant" & m_top$logFC > 0, "Positive Significant",  -->
<!--                       ifelse(m_top$Significance == "Significant" & m_top$logFC < 0, "Negative Significant",  -->
<!--                              "Non-Significant")) -->

<!-- # Select the genes that are significant to label -->
<!-- m_top$Label <- ifelse(m_top$Significance == "Significant", as.character(m_top$Gene), NA) -->

<!-- # Plot -->
<!-- p <- ggplot(m_top, aes(x = logFC, y = logPValue, color = Color)) + -->
<!--   geom_point(aes(shape = Color), size = 3, alpha = 0.7) +  # Set alpha for transparency -->
<!--   scale_color_manual(values = c("Positive Significant" = "red", "Negative Significant" = "blue", "Non-Significant" = "gray")) +  # Custom color scale -->
<!--   scale_shape_manual(values = c("Positive Significant" = 19, "Negative Significant" = 19, "Non-Significant" = 1)) +  # Same shape for all but Non-Sig -->
<!--   geom_text(aes(label = Label), size = 3, color = "black", vjust = 1, hjust = 1) +  # Add labels for significant genes in black -->
<!--   labs(x = "log2 Fold Change", y = "-log10(p-value)", title = "Volcano Plot") + -->
<!--   theme_minimal() + -->
<!--   theme(legend.position = "top") + -->
<!--   theme(axis.text = element_text(size = 12), axis.title = element_text(size = 14)) -->

<!-- # Print the plot -->
<!-- print(p) -->
<!-- #------------------------------------------------------------------------------------------------ -->

<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- length(unique(so_subgroup$kit_id)) #12 -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->


<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type_2_Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->


<!-- #Medications (AMong T2D + OB) -->
<!-- #Filter to Type_2_Diabetes only since no other participants are on these main meds -->
<!-- so_diab <- subset(so_pb,group2=="T2D + OB") -->
<!-- length(unique(so_diab$kit_id)) #35 -->
<!-- #levels = c("no_med", "sglt2", "glp1","glp1_sglt2")) -->
<!-- #1. SLGT2i+/GLP-1ra+ -->
<!-- #Combo vs. No Med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Combo vs. GLP1 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30) -->


<!-- #Combo vs. SGLT2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2") -->
<!-- length(unique(so_subgroup$kit_id)) #15 -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30) -->


<!-- #2. SGLT2i-/GLP-1ra+ -->
<!-- #GLP-1 exclusive vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30) -->

<!-- #Glp1 vs. sglt2 -->
<!-- so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30) -->

<!-- #3. SGLT2i+/GLP-1ra- -->
<!-- #sglt2 vs. no med -->
<!-- so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med") -->
<!-- so_subgroup$medication <- factor(so_subgroup$medication) -->
<!-- all_genes <- rownames(so_subgroup) -->
<!-- degs_fxn2(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30) -->







<!-- # Fit the linear mixed effects model -->
<!-- # UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV -->
<!-- vars <- c("hba1c","sbp","dbp","map","acr_u","gfr_raw_plasma","eGFR_CKD_epi","bmi","dexa_fat_kg","dexa_lean_kg","ht_adj_tkv","total_kidney_volume_ml","erpf_raw_plasma") -->
<!-- full_results <- data.frame() -->
<!-- #Zero inflated negative binomial mixed model/negative binomail mixed model -->
<!-- for (x in vars[1]) { -->
<!--   for (gene in gene_list[1:3]) { -->
<!--     # Assuming gene, x, group, and kit_id are already defined -->
<!--     data[[x]] <- as.numeric(data[[x]]) -->
<!--     m0 <- as.formula(paste0(gene," ~ ", x, "+ group + (1 | kit_id)")) -->
<!--     # model <- lmer(m0, data = data) -->
<!--     model <- glmmTMB(m0,data=data,family = nbinom2,  # Negative Binomial with a log link -->
<!--                      ziformula = ~ 1) -->
<!--     #Check if zero inflation component is non-significant, if non-sig, run nb -->
<!--     if(summary(model)$coef$zi[1,4]>=0.05) { -->
<!--       model2 <- glmmTMB(m0, data=data, -->
<!--                         family = nbinom2) -->
<!--       foldchange <- round(exp(summary(model2)$coef$cond[2,1]) - 1,5) -->
<!--       pval <- summary(model2)$coef$cond[2,4] -->
<!--       model_name <- "Negative Binomial" -->
<!--     }  -->
<!--     #If sig, pull estimates from zinb  -->
<!--     if(summary(model)$coef$zi[1,4]<0.05) { -->
<!--       foldchange <- round(exp(summary(model)$coef$cond[2,1]) - 1,5) -->
<!--       pval <- summary(model)$coef$cond[2,4] -->
<!--       model_name <- "Zero-Inflated Negative Binomial" -->
<!--     } -->
<!--     #Plot -->
<!--     p <- ggplot(data, aes(x = .data[[x]], y = .data[[gene]])) + -->
<!--       geom_point(aes(color=kit_id)) +  # Adds points to the plot -->
<!--       geom_smooth(method="lm",se=F)+ -->
<!--       labs(x = paste0(x), y = paste0(gene), -->
<!--            title=paste0(model_name," for ",gene," vs. ",str_to_title(x)), -->
<!--            subtitle=paste0("Fold Change = ",foldchange,", P-value = ",round(pval,3))) + -->
<!--       theme_minimal() + -->
<!--       theme(legend.position = "none") -->
<!--     # print(p) -->
<!--     results <- data.frame(Gene=gene,Exposure=x,FoldChange=foldchange,PValue=pval) -->
<!--     full_results <- rbind(full_results,results) -->
<!--     print(p) -->
<!--   } -->
<!-- } -->
<!-- # write.csv(full_results,fs::path(dir.results,paste0("LMEM_Results.csv"))) -->

<!-- ``` -->

<!-- #### b. PT & TAL Pseudobulk -->
<!-- ```{r, echo=F,warning=F,fig.width=15,fig.height=9} -->
<!-- #Pseuodbulk Tall and PT cells -->
<!-- so_kidney_sc$cell_type2 <- ifelse(grepl("PT-",so_kidney_sc$celltype_rpca),"PT", -->
<!--                                   ifelse(grepl("TAL-",so_kidney_sc$celltype_rpca),"TAL",NA)) -->
<!-- so_kidney_sc$cell_type2 <- as.character(so_kidney_sc$cell_type2) -->
<!-- all_cell_types <- unique(so_kidney_sc$cell_type2) -->
<!-- all_cell_types <- all_cell_types[-1] -->
<!-- Idents(so_kidney_sc) <- so_kidney_sc$cell_type2 -->

<!-- # Extract the gene expression data for all genes -->
<!-- gene_expression <- as.data.frame(so_kpmp_sc@assays$RNA@layers$data) -->

<!-- # Assign the gene names (row names of Seurat object) as the column names -->
<!-- rownames(gene_expression) <- rownames(so_kpmp_sc) -->
<!-- colnames(gene_expression) <- colnames(so_kpmp_sc) -->
<!-- gene_expression <- t(gene_expression) -->
<!-- gene_expression <- data.frame(gene_expression) -->
<!-- #log transform gene expression -->
<!-- gene_expression <- gene_expression %>%  -->
<!--   mutate(across(everything(),~log1p(.))) # log1p is log(x + 1) -->
<!-- gene_expression$cellname <- rownames(gene_expression) -->
<!-- rownames(gene_expression) <- NULL -->

<!-- # Extract the metadata, make sure you select relevant columns (e.g., 'predictor1', 'disease_status') -->
<!-- metadata <- so_kpmp_sc@meta.data -->
<!-- metadata <- metadata %>%  -->
<!--   mutate(across(everything(),~ifelse(.==".",NA,.))) -->
<!-- metadata$cellname <- rownames(metadata) -->
<!-- rownames(metadata) <- NULL -->

<!-- # filter_ids <- unique(data$kit_id[which(data[[x]]==".")]) -->

<!-- # Combine the gene expression data and metadata -->
<!-- # This ensures cells (columns) are aligned properly between the two -->
<!-- data <- tidylog::full_join(metadata,gene_expression,by="cellname") -->
<!-- # unique(data$kit_id) #81 participants -->
<!-- # which(colnames(data) =="cellname") -->
<!-- unique(data$group) -->
<!-- data$group <- factor(data$group,levels = c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes")) -->
<!-- # length(data$hba1c) -->
<!-- # length(which(data$hba1c==".")) #7457 cells missing hba1c -->
<!-- # length(unique(data$kit_id[which(data$hba1c==".")])) #3 people missing HbA1c: "KL-0029407" "KL-0030607" "KL-0031456" -->
<!-- rm(gene_expression,metadata) -->
<!-- # Fit the linear mixed effects model -->
<!-- gene_list <- colnames(gene_expression)[-9277] -->
<!-- # UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV -->
<!-- vars <- c("hba1c","sbp","dbp","map","acr_u","gfr_raw_plasma","eGFR_CKD_epi","bmi","dexa_fat_kg","dexa_lean_kg","ht_adj_tkv","total_kidney_volume_ml","erpf_raw_plasma") -->
<!-- full_results <- data.frame() -->
<!-- for (x in vars) { -->
<!--   for (gene in gene_list) { -->
<!--     data[[x]] <- as.numeric(data[[x]]) -->
<!--     m0 <- as.formula(paste0(gene," ~ ", x, "+ group + (1 | kit_id)")) -->
<!--     model <- lmer(m0, data = data) -->
<!--     foldchange <- round(exp(summary(model)$coef[2,1]) - 1,5) -->
<!--     pval <- summary(model)$coef[2,5] -->

<!--     #Plot -->
<!--     p <- ggplot(data, aes(x = .data[[x]], y = .data[[gene]])) + -->
<!--       geom_point(aes(color=kit_id)) +  # Adds points to the plot -->
<!--       geom_smooth(method="lm",se=F)+ -->
<!--       labs(x = paste0(x), y = paste0(gene), -->
<!--            title=paste0(gene," vs. ",str_to_title(x)), -->
<!--            subtitle=paste0("Fold Change = ",foldchange,", P-value = ",round(pval,3))) + -->
<!--       theme_minimal() + -->
<!--       theme(legend.position = "none") -->
<!--     # print(p) -->
<!--     results <- data.frame(Gene=gene,Exposure=x,FoldChange=foldchange,PValue=pval) -->
<!--     full_results <- rbind(full_results,results) -->
<!--     print(p) -->
<!--   } -->
<!-- } -->
<!-- write.csv(full_results,fs::path(dir.results,paste0("LMEM_Results.csv"))) -->

<!-- ``` -->

<!-- #### c. KPMP Cell Types Old -->
<!-- #### d. KPMP Cell Types New -->

<!-- ## B. Senescence -->
<!-- ## c. Metabolism -->
<!-- ## d. Inflammation -->
<!-- ## e. Growth Pathway & MTOR? -->










